<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.8.17"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>panga: panga::Genome Class Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">panga
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.17 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
var searchBox = new SearchBox("searchBox", "search",false,'Search');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */</script>
<div id="main-nav"></div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div id="nav-path" class="navpath">
  <ul>
<li class="navelem"><b>panga</b></li><li class="navelem"><a class="el" href="classpanga_1_1Genome.html">Genome</a></li>  </ul>
</div>
</div><!-- top -->
<div class="header">
  <div class="summary">
<a href="#nested-classes">Classes</a> &#124;
<a href="#pub-methods">Public Member Functions</a> &#124;
<a href="classpanga_1_1Genome-members.html">List of all members</a>  </div>
  <div class="headertitle">
<div class="title">panga::Genome Class Reference</div>  </div>
</div><!--header-->
<div class="contents">

<p><code>#include &lt;<a class="el" href="Genome_8h_source.html">Genome.h</a>&gt;</code></p>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="nested-classes"></a>
Classes</h2></td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">struct &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structpanga_1_1Genome_1_1Gene.html">Gene</a></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-methods"></a>
Public Member Functions</h2></td></tr>
<tr class="memitem:af1f4bc228d8ca6d9832ee7aa00149503"><td class="memItemLeft" align="right" valign="top"><a id="af1f4bc228d8ca6d9832ee7aa00149503"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>Genome</b> (const <a class="el" href="classpanga_1_1Genome.html">Genome</a> &amp;rhs)=delete</td></tr>
<tr class="separator:af1f4bc228d8ca6d9832ee7aa00149503"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:adf1b0433ba845a111b5e1d79ac7c9af6"><td class="memItemLeft" align="right" valign="top"><a id="adf1b0433ba845a111b5e1d79ac7c9af6"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>Genome</b> (<a class="el" href="classpanga_1_1Genome.html">Genome</a> &amp;&amp;rhs)=default</td></tr>
<tr class="separator:adf1b0433ba845a111b5e1d79ac7c9af6"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:abea2a267eefef734e880bf4368517de8"><td class="memItemLeft" align="right" valign="top"><a id="abea2a267eefef734e880bf4368517de8"></a>
<a class="el" href="classpanga_1_1Genome.html">Genome</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><b>operator=</b> (const <a class="el" href="classpanga_1_1Genome.html">Genome</a> &amp;rhs)=delete</td></tr>
<tr class="separator:abea2a267eefef734e880bf4368517de8"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a3f04e8ee6b7fe99f26a6e9031521f399"><td class="memItemLeft" align="right" valign="top">size_t&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpanga_1_1Genome.html#a3f04e8ee6b7fe99f26a6e9031521f399">AddGene</a> (size_t bit_width, bool byte_align=false)</td></tr>
<tr class="separator:a3f04e8ee6b7fe99f26a6e9031521f399"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ade8c25f66f119cdbe9fa916924139e5a"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpanga_1_1Genome.html#ade8c25f66f119cdbe9fa916924139e5a">SetBooleanGeneCount</a> (size_t boolean_gene_count)</td></tr>
<tr class="separator:ade8c25f66f119cdbe9fa916924139e5a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a1078aaf7afe7178cbbba3d931c7a17e8"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpanga_1_1Genome.html#a1078aaf7afe7178cbbba3d931c7a17e8">AddBooleanGenes</a> (size_t count)</td></tr>
<tr class="separator:a1078aaf7afe7178cbbba3d931c7a17e8"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a805ce356d2e709dc131b4772c67bd232"><td class="memItemLeft" align="right" valign="top">size_t&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpanga_1_1Genome.html#a805ce356d2e709dc131b4772c67bd232">GetBooleanGeneCount</a> () const</td></tr>
<tr class="separator:a805ce356d2e709dc131b4772c67bd232"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a17d4bb012307c43f4db5a87db9a03ee0"><td class="memItemLeft" align="right" valign="top">size_t&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpanga_1_1Genome.html#a17d4bb012307c43f4db5a87db9a03ee0">GetGeneStartBitIndex</a> (size_t gene_index) const</td></tr>
<tr class="separator:a17d4bb012307c43f4db5a87db9a03ee0"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a289b7486de854b0dce9d6b1eb5359921"><td class="memItemLeft" align="right" valign="top">size_t&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpanga_1_1Genome.html#a289b7486de854b0dce9d6b1eb5359921">GetGeneBitWitdh</a> (size_t gene_index) const</td></tr>
<tr class="separator:a289b7486de854b0dce9d6b1eb5359921"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac9377fc4051f709d160873f4b48bbe69"><td class="memItemLeft" align="right" valign="top">size_t&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpanga_1_1Genome.html#ac9377fc4051f709d160873f4b48bbe69">GetFirstBooleanGeneBitIndex</a> () const</td></tr>
<tr class="separator:ac9377fc4051f709d160873f4b48bbe69"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a2673b1bda69d21c4d5c7e5cad2783e92"><td class="memItemLeft" align="right" valign="top">size_t&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpanga_1_1Genome.html#a2673b1bda69d21c4d5c7e5cad2783e92">GetFirstBooleanGeneIndex</a> () const</td></tr>
<tr class="separator:a2673b1bda69d21c4d5c7e5cad2783e92"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a8df80cfa12ace5891c8186c1ed9f06b1"><td class="memItemLeft" align="right" valign="top">size_t&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpanga_1_1Genome.html#a8df80cfa12ace5891c8186c1ed9f06b1">GetGeneCount</a> () const</td></tr>
<tr class="separator:a8df80cfa12ace5891c8186c1ed9f06b1"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a75eb4ff4d71d5e8689336318858aac9a"><td class="memItemLeft" align="right" valign="top">size_t&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpanga_1_1Genome.html#a75eb4ff4d71d5e8689336318858aac9a">BitsRequired</a> () const</td></tr>
<tr class="separator:a75eb4ff4d71d5e8689336318858aac9a"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><p>Provides a representation of the genes which, taken together, represent a genome for members of a species.<br  />
 Using the <a class="el" href="classpanga_1_1Genome.html">Genome</a>, we may decode a block of binary data and extract the bits representing each gene from the data. (This is what <a class="el" href="classpanga_1_1Chromosome.html">Chromosome</a> does).<br  />
 The <a class="el" href="classpanga_1_1Genome.html">Genome</a> data may be constructed by adding genes one-at-a-time via AddGene along with an optional number of boolean genes.<br  />
 The genome is structured into two pieces:<br  />
</p><ol type="1">
<li>Genes with a bit width<br  />
</li>
<li>Boolean genes which have only one bit value<br  />
 For performance reasons, all of the boolean genes are located at the end of the <a class="el" href="classpanga_1_1Genome.html">Genome</a>. It is preferable to add a boolean gene than add a gene with bit width of 1. <dl class="section see"><dt>See also</dt><dd><a class="el" href="classpanga_1_1Genome.html#a3f04e8ee6b7fe99f26a6e9031521f399">AddGene</a> </dd>
<dd>
<a class="el" href="classpanga_1_1Chromosome.html">Chromosome</a> </dd></dl>
</li>
</ol>
</div><h2 class="groupheader">Member Function Documentation</h2>
<a id="a1078aaf7afe7178cbbba3d931c7a17e8"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a1078aaf7afe7178cbbba3d931c7a17e8">&#9670;&nbsp;</a></span>AddBooleanGenes()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void panga::Genome::AddBooleanGenes </td>
          <td>(</td>
          <td class="paramtype">size_t&#160;</td>
          <td class="paramname"><em>count</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Add count boolean genes to the boolean gene count for this <a class="el" href="classpanga_1_1Genome.html">Genome</a>.<br  />
 Does not return the gene indices because adding more non-boolean genes to the <a class="el" href="classpanga_1_1Genome.html">Genome</a> can change the boolean gene indices. </p><dl class="section see"><dt>See also</dt><dd><a class="el" href="classpanga_1_1Genome.html#a2673b1bda69d21c4d5c7e5cad2783e92">GetFirstBooleanGeneIndex</a> </dd>
<dd>
<a class="el" href="classpanga_1_1Genome.html#ac9377fc4051f709d160873f4b48bbe69">GetFirstBooleanGeneBitIndex</a> </dd></dl>

</div>
</div>
<a id="a3f04e8ee6b7fe99f26a6e9031521f399"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a3f04e8ee6b7fe99f26a6e9031521f399">&#9670;&nbsp;</a></span>AddGene()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">size_t panga::Genome::AddGene </td>
          <td>(</td>
          <td class="paramtype">size_t&#160;</td>
          <td class="paramname"><em>bit_width</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">bool&#160;</td>
          <td class="paramname"><em>byte_align</em> = <code>false</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Add a gene to the <a class="el" href="classpanga_1_1Genome.html">Genome</a>.<br  />
 The new gene will be added to the end of the <a class="el" href="classpanga_1_1Genome.html">Genome</a>.<br  />
 This can be very useful if you want to access the memory backing this gene later directly (via something like <a class="el" href="classpanga_1_1Chromosome.html#aa0401680be20878119042d33741068b1">Chromosome::GetRawGene</a>). </p><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">bit_width</td><td>The width of the gene we want to add. </td></tr>
    <tr><td class="paramname">byte_align</td><td>If true, we will shift the starting offset and bit width such that this gene begins and ends on a byte-boundary. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="classpanga_1_1Chromosome.html#aa0401680be20878119042d33741068b1">Chromosome::GetRawGene</a> </dd></dl>
<dl class="section return"><dt>Returns</dt><dd>The gene index of the newly-added gene. </dd></dl>

</div>
</div>
<a id="a75eb4ff4d71d5e8689336318858aac9a"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a75eb4ff4d71d5e8689336318858aac9a">&#9670;&nbsp;</a></span>BitsRequired()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">size_t panga::Genome::BitsRequired </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Get the number of bits needed to encode this <a class="el" href="classpanga_1_1Genome.html">Genome</a>. </p>

</div>
</div>
<a id="a805ce356d2e709dc131b4772c67bd232"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a805ce356d2e709dc131b4772c67bd232">&#9670;&nbsp;</a></span>GetBooleanGeneCount()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">size_t panga::Genome::GetBooleanGeneCount </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Get the number of boolean Genes in the <a class="el" href="classpanga_1_1Genome.html">Genome</a>. </p>

</div>
</div>
<a id="ac9377fc4051f709d160873f4b48bbe69"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ac9377fc4051f709d160873f4b48bbe69">&#9670;&nbsp;</a></span>GetFirstBooleanGeneBitIndex()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">size_t panga::Genome::GetFirstBooleanGeneBitIndex </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Get the bit index of the first boolean gene in the <a class="el" href="classpanga_1_1Genome.html">Genome</a>. </p>

</div>
</div>
<a id="a2673b1bda69d21c4d5c7e5cad2783e92"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a2673b1bda69d21c4d5c7e5cad2783e92">&#9670;&nbsp;</a></span>GetFirstBooleanGeneIndex()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">size_t panga::Genome::GetFirstBooleanGeneIndex </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Get the gene index of the first boolean gene in the <a class="el" href="classpanga_1_1Genome.html">Genome</a>. </p>

</div>
</div>
<a id="a289b7486de854b0dce9d6b1eb5359921"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a289b7486de854b0dce9d6b1eb5359921">&#9670;&nbsp;</a></span>GetGeneBitWitdh()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">size_t panga::Genome::GetGeneBitWitdh </td>
          <td>(</td>
          <td class="paramtype">size_t&#160;</td>
          <td class="paramname"><em>gene_index</em></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Get the bit width for a gene. </p><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">gene_index</td><td>Index of the gene we should return the width for. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>The width in bits of gene at index gene_index. </dd></dl>

</div>
</div>
<a id="a8df80cfa12ace5891c8186c1ed9f06b1"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a8df80cfa12ace5891c8186c1ed9f06b1">&#9670;&nbsp;</a></span>GetGeneCount()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">size_t panga::Genome::GetGeneCount </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Get the count of genes in the <a class="el" href="classpanga_1_1Genome.html">Genome</a>.<br  />
 Note: Includes the count of boolean genes. </p>

</div>
</div>
<a id="a17d4bb012307c43f4db5a87db9a03ee0"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a17d4bb012307c43f4db5a87db9a03ee0">&#9670;&nbsp;</a></span>GetGeneStartBitIndex()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">size_t panga::Genome::GetGeneStartBitIndex </td>
          <td>(</td>
          <td class="paramtype">size_t&#160;</td>
          <td class="paramname"><em>gene_index</em></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Get the bit offset where a gene starts. </p>

</div>
</div>
<a id="ade8c25f66f119cdbe9fa916924139e5a"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ade8c25f66f119cdbe9fa916924139e5a">&#9670;&nbsp;</a></span>SetBooleanGeneCount()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void panga::Genome::SetBooleanGeneCount </td>
          <td>(</td>
          <td class="paramtype">size_t&#160;</td>
          <td class="paramname"><em>boolean_gene_count</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Set the number of boolean genes in this <a class="el" href="classpanga_1_1Genome.html">Genome</a>.<br  />
 This overwrites any value currently in the boolean gene count. </p>

</div>
</div>
<hr/>The documentation for this class was generated from the following files:<ul>
<li>src/<a class="el" href="Genome_8h_source.html">Genome.h</a></li>
<li>src/Genome.cc</li>
</ul>
</div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by &#160;<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/>
</a> 1.8.17
</small></address>
</body>
</html>
