<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.8.13"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>panga: panga::GeneticAlgorithm Class Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">panga
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.13 -->
<script type="text/javascript">
var searchBox = new SearchBox("searchBox", "search",false,'Search');
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
</script>
<div id="main-nav"></div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div id="nav-path" class="navpath">
  <ul>
<li class="navelem"><b>panga</b></li><li class="navelem"><a class="el" href="classpanga_1_1GeneticAlgorithm.html">GeneticAlgorithm</a></li>  </ul>
</div>
</div><!-- top -->
<div class="header">
  <div class="summary">
<a href="#pub-types">Public Types</a> &#124;
<a href="#pub-methods">Public Member Functions</a> &#124;
<a href="#pro-methods">Protected Member Functions</a> &#124;
<a href="#pro-attribs">Protected Attributes</a> &#124;
<a href="classpanga_1_1GeneticAlgorithm-members.html">List of all members</a>  </div>
  <div class="headertitle">
<div class="title">panga::GeneticAlgorithm Class Reference</div>  </div>
</div><!--header-->
<div class="contents">
<div class="dynheader">
Collaboration diagram for panga::GeneticAlgorithm:</div>
<div class="dyncontent">
<div class="center"><img src="classpanga_1_1GeneticAlgorithm__coll__graph.png" border="0" usemap="#panga_1_1GeneticAlgorithm_coll__map" alt="Collaboration graph"/></div>
<map name="panga_1_1GeneticAlgorithm_coll__map" id="panga_1_1GeneticAlgorithm_coll__map">
<area shape="rect" id="node2" href="classpanga_1_1Population.html" title="panga::Population" alt="" coords="34,184,173,211"/>
<area shape="rect" id="node4" href="classpanga_1_1Genome.html" title="panga::Genome" alt="" coords="238,5,363,32"/>
<area shape="rect" id="node6" href="classpanga_1_1Chromosome.html" title="panga::Chromosome" alt="" coords="338,95,495,121"/>
<area shape="rect" id="node5" href="classpanga_1_1Individual.html" title="panga::Individual" alt="" coords="339,184,473,211"/>
<area shape="rect" id="node7" href="classpanga_1_1BitVector.html" title="panga::BitVector" alt="" coords="388,5,517,32"/>
<area shape="rect" id="node8" href="classpanga_1_1RandomWrapper.html" title="panga::RandomWrapper" alt="" coords="498,184,677,211"/>
</map>
<center><span class="legend">[<a href="graph_legend.html">legend</a>]</span></center></div>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-types"></a>
Public Types</h2></td></tr>
<tr class="memitem:a121354609fbc300abfb96556935f4b35"><td class="memItemLeft" align="right" valign="top">enum &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpanga_1_1GeneticAlgorithm.html#a121354609fbc300abfb96556935f4b35">CrossoverType</a> : uint8_t { <a class="el" href="classpanga_1_1GeneticAlgorithm.html#a121354609fbc300abfb96556935f4b35af19d8dfb9f04573bad667420b8cb2270">CrossoverType::OnePoint</a> = 1, 
<a class="el" href="classpanga_1_1GeneticAlgorithm.html#a121354609fbc300abfb96556935f4b35a331d16a2885c7f82ba949f286f48143d">CrossoverType::TwoPoint</a>, 
<a class="el" href="classpanga_1_1GeneticAlgorithm.html#a121354609fbc300abfb96556935f4b35a0d398d8c2d3f38c8884ae132c394b784">CrossoverType::KPoint</a>, 
<a class="el" href="classpanga_1_1GeneticAlgorithm.html#a121354609fbc300abfb96556935f4b35af19516d11f2946f894070e92fcb56b6d">CrossoverType::Uniform</a>
 }</td></tr>
<tr class="separator:a121354609fbc300abfb96556935f4b35"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a175a0790ea362ea9794fa4daa222d818"><td class="memItemLeft" align="right" valign="top">enum &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpanga_1_1GeneticAlgorithm.html#a175a0790ea362ea9794fa4daa222d818">MutatorType</a> : uint8_t { <a class="el" href="classpanga_1_1GeneticAlgorithm.html#a175a0790ea362ea9794fa4daa222d818a9ffbd422925a6839ee820ddbc59278c5">MutatorType::Flip</a> = 1
 }</td></tr>
<tr class="separator:a175a0790ea362ea9794fa4daa222d818"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a3653afdb62876d5d18254dba4de27a6e"><td class="memItemLeft" align="right" valign="top">enum &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpanga_1_1GeneticAlgorithm.html#a3653afdb62876d5d18254dba4de27a6e">SelectorType</a> : uint8_t { <a class="el" href="classpanga_1_1GeneticAlgorithm.html#a3653afdb62876d5d18254dba4de27a6ea021da1b20f73dc252361a54d80497ef3">SelectorType::Rank</a> = 1, 
<a class="el" href="classpanga_1_1GeneticAlgorithm.html#a3653afdb62876d5d18254dba4de27a6eaf19516d11f2946f894070e92fcb56b6d">SelectorType::Uniform</a>, 
<a class="el" href="classpanga_1_1GeneticAlgorithm.html#a3653afdb62876d5d18254dba4de27a6eaf1c787f267b4d6d03f1202063099a24a">SelectorType::RouletteWheel</a>, 
<a class="el" href="classpanga_1_1GeneticAlgorithm.html#a3653afdb62876d5d18254dba4de27a6ea4f4dc29df9d29cbec61a725ff3ce6e72">SelectorType::Tournament</a>
 }</td></tr>
<tr class="separator:a3653afdb62876d5d18254dba4de27a6e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a84ecf735778d44eceefdc7c9f5a4f21b"><td class="memItemLeft" align="right" valign="top">enum &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpanga_1_1GeneticAlgorithm.html#a84ecf735778d44eceefdc7c9f5a4f21b">MutationRateSchedule</a> : uint8_t { <a class="el" href="classpanga_1_1GeneticAlgorithm.html#a84ecf735778d44eceefdc7c9f5a4f21bacb17869fe51048b5a5c4c6106551a255">MutationRateSchedule::Constant</a> = 1, 
<a class="el" href="classpanga_1_1GeneticAlgorithm.html#a84ecf735778d44eceefdc7c9f5a4f21bac172d0b7b7d4c61a931d50369d37b11f">MutationRateSchedule::Deterministic</a>, 
<a class="el" href="classpanga_1_1GeneticAlgorithm.html#a84ecf735778d44eceefdc7c9f5a4f21baf5f333dcdae61f0aa41cd7b8c82222aa">MutationRateSchedule::SelfAdaptive</a>, 
<a class="el" href="classpanga_1_1GeneticAlgorithm.html#a84ecf735778d44eceefdc7c9f5a4f21baac13b33ebc8ea372dfb07e5ed4d9655f">MutationRateSchedule::Proportional</a>
 }</td></tr>
<tr class="separator:a84ecf735778d44eceefdc7c9f5a4f21b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:adacbd6c9491bc61df372c0697b758a99"><td class="memItemLeft" align="right" valign="top"><a id="adacbd6c9491bc61df372c0697b758a99"></a>
typedef double(*&#160;</td><td class="memItemRight" valign="bottom"><b>FitnessFunction</b>) (<a class="el" href="classpanga_1_1Individual.html">Individual</a> *, void *)</td></tr>
<tr class="separator:adacbd6c9491bc61df372c0697b758a99"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-methods"></a>
Public Member Functions</h2></td></tr>
<tr class="memitem:aeaf71c63931a61074b12d73f995e8699"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpanga_1_1GeneticAlgorithm.html#aeaf71c63931a61074b12d73f995e8699">SetGenome</a> (const <a class="el" href="classpanga_1_1Genome.html">Genome</a> *genome)</td></tr>
<tr class="separator:aeaf71c63931a61074b12d73f995e8699"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a32864ad611c469cf52d19d36c8e865ff"><td class="memItemLeft" align="right" valign="top"><a id="a32864ad611c469cf52d19d36c8e865ff"></a>
const <a class="el" href="classpanga_1_1Genome.html">Genome</a> *&#160;</td><td class="memItemRight" valign="bottom"><b>GetGenome</b> ()</td></tr>
<tr class="separator:a32864ad611c469cf52d19d36c8e865ff"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af5db5af0bac1cd2923cfc8dab414ba59"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpanga_1_1GeneticAlgorithm.html#af5db5af0bac1cd2923cfc8dab414ba59">SetCrossoverIgnoreGeneBoundaries</a> (bool crossoverIgnoreGeneBoundaries)</td></tr>
<tr class="separator:af5db5af0bac1cd2923cfc8dab414ba59"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a0f8724876a3ad7c9df9e20bfb083fd5e"><td class="memItemLeft" align="right" valign="top"><a id="a0f8724876a3ad7c9df9e20bfb083fd5e"></a>
bool&#160;</td><td class="memItemRight" valign="bottom"><b>GetCrossoverIgnoreGeneBoundaries</b> ()</td></tr>
<tr class="separator:a0f8724876a3ad7c9df9e20bfb083fd5e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a56014a62f4853e76701abbdc3db0cb18"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpanga_1_1GeneticAlgorithm.html#a56014a62f4853e76701abbdc3db0cb18">SetAllowSameParentCouples</a> (bool allowSameParentCouples)</td></tr>
<tr class="separator:a56014a62f4853e76701abbdc3db0cb18"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a5ddef8642ec08dc6f70ff595a10e7de2"><td class="memItemLeft" align="right" valign="top"><a id="a5ddef8642ec08dc6f70ff595a10e7de2"></a>
bool&#160;</td><td class="memItemRight" valign="bottom"><b>GetAllowSameParentCouples</b> ()</td></tr>
<tr class="separator:a5ddef8642ec08dc6f70ff595a10e7de2"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad6deca0de2a42a646a5d2b48f83a12b6"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpanga_1_1GeneticAlgorithm.html#ad6deca0de2a42a646a5d2b48f83a12b6">SetTournamentSize</a> (size_t tournamentSize)</td></tr>
<tr class="separator:ad6deca0de2a42a646a5d2b48f83a12b6"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad9c4e0d1dc6a51bd0b3f81eab6e86540"><td class="memItemLeft" align="right" valign="top"><a id="ad9c4e0d1dc6a51bd0b3f81eab6e86540"></a>
size_t&#160;</td><td class="memItemRight" valign="bottom"><b>GetTournamentSize</b> ()</td></tr>
<tr class="separator:ad9c4e0d1dc6a51bd0b3f81eab6e86540"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a72a9de6fcb54fc424a4fa5d66f388907"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpanga_1_1GeneticAlgorithm.html#a72a9de6fcb54fc424a4fa5d66f388907">SetKPointCrossoverPointCount</a> (size_t kPointCrossoverPointCount)</td></tr>
<tr class="separator:a72a9de6fcb54fc424a4fa5d66f388907"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a2564c514efb9730673509d83e8470848"><td class="memItemLeft" align="right" valign="top"><a id="a2564c514efb9730673509d83e8470848"></a>
size_t&#160;</td><td class="memItemRight" valign="bottom"><b>GetKPointCrossoverPointCount</b> ()</td></tr>
<tr class="separator:a2564c514efb9730673509d83e8470848"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aca07c13c4b683605270cb2f0ea24a28c"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpanga_1_1GeneticAlgorithm.html#aca07c13c4b683605270cb2f0ea24a28c">SetSelfAdaptiveMutationDiversityFloor</a> (double selfAdaptiveMutationDiversityFloor)</td></tr>
<tr class="separator:aca07c13c4b683605270cb2f0ea24a28c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a196907cc18c8d5ff30945098b586eb3b"><td class="memItemLeft" align="right" valign="top"><a id="a196907cc18c8d5ff30945098b586eb3b"></a>
double&#160;</td><td class="memItemRight" valign="bottom"><b>GetSelfAdaptiveMutationDiversityFloor</b> ()</td></tr>
<tr class="separator:a196907cc18c8d5ff30945098b586eb3b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a07a9e112cb63a3a6ddc101a3a31e0411"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpanga_1_1GeneticAlgorithm.html#a07a9e112cb63a3a6ddc101a3a31e0411">SetSelfAdaptiveMutationAggressiveRate</a> (double selfAdaptiveMutationAggressiveRate)</td></tr>
<tr class="separator:a07a9e112cb63a3a6ddc101a3a31e0411"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae430d242f31e77583fa40530bdc77135"><td class="memItemLeft" align="right" valign="top"><a id="ae430d242f31e77583fa40530bdc77135"></a>
double&#160;</td><td class="memItemRight" valign="bottom"><b>GetSelfAdaptiveMutationAggressiveRate</b> ()</td></tr>
<tr class="separator:ae430d242f31e77583fa40530bdc77135"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a536b029dd25bdd78a17fcbefd21f8b11"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpanga_1_1GeneticAlgorithm.html#a536b029dd25bdd78a17fcbefd21f8b11">SetProportionalMutationBitCount</a> (size_t proportionalMutationBitCount)</td></tr>
<tr class="separator:a536b029dd25bdd78a17fcbefd21f8b11"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a09f5ccd39b81a29cf718508c8a208640"><td class="memItemLeft" align="right" valign="top"><a id="a09f5ccd39b81a29cf718508c8a208640"></a>
size_t&#160;</td><td class="memItemRight" valign="bottom"><b>GetProportionalMutationBitCount</b> ()</td></tr>
<tr class="separator:a09f5ccd39b81a29cf718508c8a208640"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:add6b138984b9541983cad01774a2b3c9"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpanga_1_1GeneticAlgorithm.html#add6b138984b9541983cad01774a2b3c9">SetTotalGenerations</a> (size_t totalGenerations)</td></tr>
<tr class="separator:add6b138984b9541983cad01774a2b3c9"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a33d0ccf14fea9e125185c05866fd7a29"><td class="memItemLeft" align="right" valign="top"><a id="a33d0ccf14fea9e125185c05866fd7a29"></a>
size_t&#160;</td><td class="memItemRight" valign="bottom"><b>GetTotalGenerations</b> ()</td></tr>
<tr class="separator:a33d0ccf14fea9e125185c05866fd7a29"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a4b0b1a82ef3d8e6881927ca96b27947c"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpanga_1_1GeneticAlgorithm.html#a4b0b1a82ef3d8e6881927ca96b27947c">SetMutationRate</a> (double mutationRate)</td></tr>
<tr class="separator:a4b0b1a82ef3d8e6881927ca96b27947c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa83d8fe0cd1fcaae9eb544219e36d2a7"><td class="memItemLeft" align="right" valign="top"><a id="aa83d8fe0cd1fcaae9eb544219e36d2a7"></a>
double&#160;</td><td class="memItemRight" valign="bottom"><b>GetMutationRate</b> ()</td></tr>
<tr class="separator:aa83d8fe0cd1fcaae9eb544219e36d2a7"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a4a3351748f07e44a99e86a6ac0f85f27"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpanga_1_1GeneticAlgorithm.html#a4a3351748f07e44a99e86a6ac0f85f27">SetCrossoverRate</a> (double crossoverRate)</td></tr>
<tr class="separator:a4a3351748f07e44a99e86a6ac0f85f27"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab10868329434a72661bb5c510b467030"><td class="memItemLeft" align="right" valign="top"><a id="ab10868329434a72661bb5c510b467030"></a>
double&#160;</td><td class="memItemRight" valign="bottom"><b>GetCrossoverRate</b> ()</td></tr>
<tr class="separator:ab10868329434a72661bb5c510b467030"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a070d1761aaa17dd6554cc23a0dc34f97"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpanga_1_1GeneticAlgorithm.html#a070d1761aaa17dd6554cc23a0dc34f97">SetCrossoverType</a> (<a class="el" href="classpanga_1_1GeneticAlgorithm.html#a121354609fbc300abfb96556935f4b35">CrossoverType</a> crossoverType)</td></tr>
<tr class="separator:a070d1761aaa17dd6554cc23a0dc34f97"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a3e1e0a16c003c91e1807e340e9a08c53"><td class="memItemLeft" align="right" valign="top"><a id="a3e1e0a16c003c91e1807e340e9a08c53"></a>
<a class="el" href="classpanga_1_1GeneticAlgorithm.html#a121354609fbc300abfb96556935f4b35">CrossoverType</a>&#160;</td><td class="memItemRight" valign="bottom"><b>GetCrossoverType</b> ()</td></tr>
<tr class="separator:a3e1e0a16c003c91e1807e340e9a08c53"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab0764a6c0e0db8cd6a82d109a9789e9a"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpanga_1_1GeneticAlgorithm.html#ab0764a6c0e0db8cd6a82d109a9789e9a">SetSelectorType</a> (<a class="el" href="classpanga_1_1GeneticAlgorithm.html#a3653afdb62876d5d18254dba4de27a6e">SelectorType</a> selectorType)</td></tr>
<tr class="separator:ab0764a6c0e0db8cd6a82d109a9789e9a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab0182e7002825b5f2566944ddc295872"><td class="memItemLeft" align="right" valign="top"><a id="ab0182e7002825b5f2566944ddc295872"></a>
<a class="el" href="classpanga_1_1GeneticAlgorithm.html#a3653afdb62876d5d18254dba4de27a6e">SelectorType</a>&#160;</td><td class="memItemRight" valign="bottom"><b>GetSelectorType</b> ()</td></tr>
<tr class="separator:ab0182e7002825b5f2566944ddc295872"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a9df556f467a9884b449708e03316b302"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpanga_1_1GeneticAlgorithm.html#a9df556f467a9884b449708e03316b302">SetEliteCount</a> (size_t eliteCount)</td></tr>
<tr class="separator:a9df556f467a9884b449708e03316b302"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a1b6869ea6f815a11197952d750c54a13"><td class="memItemLeft" align="right" valign="top"><a id="a1b6869ea6f815a11197952d750c54a13"></a>
size_t&#160;</td><td class="memItemRight" valign="bottom"><b>GetEliteCount</b> ()</td></tr>
<tr class="separator:a1b6869ea6f815a11197952d750c54a13"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa1c0bbfc523f2b72fa7af30fc2a95641"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpanga_1_1GeneticAlgorithm.html#aa1c0bbfc523f2b72fa7af30fc2a95641">SetMutatedEliteCount</a> (size_t mutatedEliteCount)</td></tr>
<tr class="separator:aa1c0bbfc523f2b72fa7af30fc2a95641"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af48a967d79a6fb9210235f73b658de7e"><td class="memItemLeft" align="right" valign="top"><a id="af48a967d79a6fb9210235f73b658de7e"></a>
size_t&#160;</td><td class="memItemRight" valign="bottom"><b>GetMutatedEliteCount</b> ()</td></tr>
<tr class="separator:af48a967d79a6fb9210235f73b658de7e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a4a8bd65c3dd58a49792ab370e398236b"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpanga_1_1GeneticAlgorithm.html#a4a8bd65c3dd58a49792ab370e398236b">SetMutatedEliteMutationRate</a> (double mutatedEliteMutationRate)</td></tr>
<tr class="separator:a4a8bd65c3dd58a49792ab370e398236b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a73d65f4ba06ecda8307d48ab6f515c01"><td class="memItemLeft" align="right" valign="top"><a id="a73d65f4ba06ecda8307d48ab6f515c01"></a>
double&#160;</td><td class="memItemRight" valign="bottom"><b>GetMutatedEliteMutationRate</b> ()</td></tr>
<tr class="separator:a73d65f4ba06ecda8307d48ab6f515c01"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a3e3cb4264ee3ae4a892dbd6a2bf43276"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpanga_1_1GeneticAlgorithm.html#a3e3cb4264ee3ae4a892dbd6a2bf43276">SetMutationRateSchedule</a> (<a class="el" href="classpanga_1_1GeneticAlgorithm.html#a84ecf735778d44eceefdc7c9f5a4f21b">MutationRateSchedule</a> mutationRateSchedule)</td></tr>
<tr class="separator:a3e3cb4264ee3ae4a892dbd6a2bf43276"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a21439ddac5535f5df37726671594cfce"><td class="memItemLeft" align="right" valign="top"><a id="a21439ddac5535f5df37726671594cfce"></a>
<a class="el" href="classpanga_1_1GeneticAlgorithm.html#a84ecf735778d44eceefdc7c9f5a4f21b">MutationRateSchedule</a>&#160;</td><td class="memItemRight" valign="bottom"><b>GetMutationRateSchedule</b> ()</td></tr>
<tr class="separator:a21439ddac5535f5df37726671594cfce"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae43cef679831a1917d13795401387250"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpanga_1_1GeneticAlgorithm.html#ae43cef679831a1917d13795401387250">SetMutatorType</a> (<a class="el" href="classpanga_1_1GeneticAlgorithm.html#a175a0790ea362ea9794fa4daa222d818">MutatorType</a> mutatorType)</td></tr>
<tr class="separator:ae43cef679831a1917d13795401387250"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a12fd9254363f6cddd57dcafea99d253f"><td class="memItemLeft" align="right" valign="top"><a id="a12fd9254363f6cddd57dcafea99d253f"></a>
<a class="el" href="classpanga_1_1GeneticAlgorithm.html#a175a0790ea362ea9794fa4daa222d818">MutatorType</a>&#160;</td><td class="memItemRight" valign="bottom"><b>GetMutatorType</b> ()</td></tr>
<tr class="separator:a12fd9254363f6cddd57dcafea99d253f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af2f1560069922de9478b57e6d6960b93"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpanga_1_1GeneticAlgorithm.html#af2f1560069922de9478b57e6d6960b93">SetUserData</a> (void *userData)</td></tr>
<tr class="separator:af2f1560069922de9478b57e6d6960b93"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a59fd2d06a4db66581b4e5f2324e2d3c8"><td class="memItemLeft" align="right" valign="top"><a id="a59fd2d06a4db66581b4e5f2324e2d3c8"></a>
void *&#160;</td><td class="memItemRight" valign="bottom"><b>GetUserData</b> ()</td></tr>
<tr class="separator:a59fd2d06a4db66581b4e5f2324e2d3c8"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a92d0aa5d3796aca95e84897323328667"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpanga_1_1GeneticAlgorithm.html#a92d0aa5d3796aca95e84897323328667">SetFitnessFunction</a> (FitnessFunction fitnessFunction)</td></tr>
<tr class="separator:a92d0aa5d3796aca95e84897323328667"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aaa92c009bfb8f03f9ac0bac3fee5c44f"><td class="memItemLeft" align="right" valign="top"><a id="aaa92c009bfb8f03f9ac0bac3fee5c44f"></a>
FitnessFunction&#160;</td><td class="memItemRight" valign="bottom"><b>GetFitnessFunction</b> ()</td></tr>
<tr class="separator:aaa92c009bfb8f03f9ac0bac3fee5c44f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae1f5aad716fae81c066428b6f2069d25"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpanga_1_1GeneticAlgorithm.html#ae1f5aad716fae81c066428b6f2069d25">SetPopulationSize</a> (size_t populationSize)</td></tr>
<tr class="separator:ae1f5aad716fae81c066428b6f2069d25"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab045e6356ee2db4a83c604c243b11d83"><td class="memItemLeft" align="right" valign="top"><a id="ab045e6356ee2db4a83c604c243b11d83"></a>
size_t&#160;</td><td class="memItemRight" valign="bottom"><b>GetPopulationSize</b> ()</td></tr>
<tr class="separator:ab045e6356ee2db4a83c604c243b11d83"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a40cbc10412d0bcc71f1b9bc2a6d1ebaf"><td class="memItemLeft" align="right" valign="top">size_t&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpanga_1_1GeneticAlgorithm.html#a40cbc10412d0bcc71f1b9bc2a6d1ebaf">GetCurrentGeneration</a> ()</td></tr>
<tr class="separator:a40cbc10412d0bcc71f1b9bc2a6d1ebaf"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aaf9685a149c8c2102e9ccc54c43d09b4"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classpanga_1_1Individual.html">Individual</a> *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpanga_1_1GeneticAlgorithm.html#aaf9685a149c8c2102e9ccc54c43d09b4">GetBestIndividual</a> ()</td></tr>
<tr class="separator:aaf9685a149c8c2102e9ccc54c43d09b4"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a0791b5149f23eb2c397592d8db5ebe19"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classpanga_1_1Individual.html">Individual</a> *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpanga_1_1GeneticAlgorithm.html#a0791b5149f23eb2c397592d8db5ebe19">GetIndividual</a> (size_t index)</td></tr>
<tr class="separator:a0791b5149f23eb2c397592d8db5ebe19"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af3c2d5f24c373871614aa2ff80c07289"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpanga_1_1GeneticAlgorithm.html#af3c2d5f24c373871614aa2ff80c07289">GetMinimumScore</a> ()</td></tr>
<tr class="separator:af3c2d5f24c373871614aa2ff80c07289"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a78e68efaff7967fb5937b680f3966353"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpanga_1_1GeneticAlgorithm.html#a78e68efaff7967fb5937b680f3966353">GetAverageScore</a> ()</td></tr>
<tr class="separator:a78e68efaff7967fb5937b680f3966353"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:adcc513420e5c5808442c1d9c84932a22"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpanga_1_1GeneticAlgorithm.html#adcc513420e5c5808442c1d9c84932a22">GetScoreStandardDeviation</a> ()</td></tr>
<tr class="separator:adcc513420e5c5808442c1d9c84932a22"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a573c741a7c511c58ca1ef0fad1935ee1"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpanga_1_1GeneticAlgorithm.html#a573c741a7c511c58ca1ef0fad1935ee1">GetPopulationDiversity</a> ()</td></tr>
<tr class="separator:a573c741a7c511c58ca1ef0fad1935ee1"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af5317922e4f09dbe17a2f59b9dd615c2"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpanga_1_1GeneticAlgorithm.html#af5317922e4f09dbe17a2f59b9dd615c2">Initialize</a> (const std::vector&lt; const <a class="el" href="classpanga_1_1BitVector.html">BitVector</a> *&gt; *initialPopulation=nullptr)</td></tr>
<tr class="separator:af5317922e4f09dbe17a2f59b9dd615c2"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a86aaa0a83bc03d318d0c6a4d30c9dbeb"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpanga_1_1GeneticAlgorithm.html#a86aaa0a83bc03d318d0c6a4d30c9dbeb">Step</a> ()</td></tr>
<tr class="separator:a86aaa0a83bc03d318d0c6a4d30c9dbeb"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aac0e857b86993a430331bbff5299b231"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpanga_1_1GeneticAlgorithm.html#aac0e857b86993a430331bbff5299b231">Run</a> ()</td></tr>
<tr class="separator:aac0e857b86993a430331bbff5299b231"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pro-methods"></a>
Protected Member Functions</h2></td></tr>
<tr class="memitem:ab4dceb9c006de4994b1e1d6f13032e83"><td class="memItemLeft" align="right" valign="top"><a id="ab4dceb9c006de4994b1e1d6f13032e83"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>GeneticAlgorithm</b> (const <a class="el" href="classpanga_1_1GeneticAlgorithm.html">GeneticAlgorithm</a> &amp;)</td></tr>
<tr class="separator:ab4dceb9c006de4994b1e1d6f13032e83"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a71413b202033cabb4b2c12e2a9e0e04a"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpanga_1_1GeneticAlgorithm.html#a71413b202033cabb4b2c12e2a9e0e04a">DeletePopulation</a> (<a class="el" href="classpanga_1_1Population.html">Population</a> *population)</td></tr>
<tr class="separator:a71413b202033cabb4b2c12e2a9e0e04a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a36729bf431bb5887ba64275754e09723"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpanga_1_1GeneticAlgorithm.html#a36729bf431bb5887ba64275754e09723">Evaluate</a> (<a class="el" href="classpanga_1_1Population.html">Population</a> *population)</td></tr>
<tr class="separator:a36729bf431bb5887ba64275754e09723"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a08b90a76a83015c56a37b9a2cff28129"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpanga_1_1GeneticAlgorithm.html#a08b90a76a83015c56a37b9a2cff28129">Score</a> (<a class="el" href="classpanga_1_1Individual.html">Individual</a> *individual)</td></tr>
<tr class="separator:a08b90a76a83015c56a37b9a2cff28129"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a511a723e317952a68e5037acf9e7a2f0"><td class="memTemplParams" colspan="2"><a id="a511a723e317952a68e5037acf9e7a2f0"></a>
template&lt;bool ignoreGeneBoundaries&gt; </td></tr>
<tr class="memitem:a511a723e317952a68e5037acf9e7a2f0"><td class="memTemplItemLeft" align="right" valign="top">void&#160;</td><td class="memTemplItemRight" valign="bottom"><b>CrossoverInternal</b> (<a class="el" href="classpanga_1_1Individual.html">Individual</a> *parent1, <a class="el" href="classpanga_1_1Individual.html">Individual</a> *parent2, <a class="el" href="classpanga_1_1Individual.html">Individual</a> *offspring)</td></tr>
<tr class="separator:a511a723e317952a68e5037acf9e7a2f0"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a7f87def6756eec276bf6456778d89d94"><td class="memItemLeft" align="right" valign="top"><a id="a7f87def6756eec276bf6456778d89d94"></a>
void&#160;</td><td class="memItemRight" valign="bottom"><b>Crossover</b> (<a class="el" href="classpanga_1_1Individual.html">Individual</a> *parent1, <a class="el" href="classpanga_1_1Individual.html">Individual</a> *parent2, <a class="el" href="classpanga_1_1Individual.html">Individual</a> *offspring)</td></tr>
<tr class="separator:a7f87def6756eec276bf6456778d89d94"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a61a4892ef8855c8202538469bc8cd178"><td class="memItemLeft" align="right" valign="top"><a id="a61a4892ef8855c8202538469bc8cd178"></a>
void&#160;</td><td class="memItemRight" valign="bottom"><b>Mutate</b> (<a class="el" href="classpanga_1_1Individual.html">Individual</a> *individual, double mutationPercentage)</td></tr>
<tr class="separator:a61a4892ef8855c8202538469bc8cd178"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a69955a35ef4429ea26382acd4cb44915"><td class="memItemLeft" align="right" valign="top"><a id="a69955a35ef4429ea26382acd4cb44915"></a>
double&#160;</td><td class="memItemRight" valign="bottom"><b>GetCurrentMutationRate</b> ()</td></tr>
<tr class="separator:a69955a35ef4429ea26382acd4cb44915"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a7bddff15089472135cdb201e4a8e73a2"><td class="memItemLeft" align="right" valign="top"><a id="a7bddff15089472135cdb201e4a8e73a2"></a>
void&#160;</td><td class="memItemRight" valign="bottom"><b>InitializeSelector</b> (<a class="el" href="classpanga_1_1Population.html">Population</a> *population)</td></tr>
<tr class="separator:a7bddff15089472135cdb201e4a8e73a2"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a1042a5efa8486919bba28548534069da"><td class="memItemLeft" align="right" valign="top"><a id="a1042a5efa8486919bba28548534069da"></a>
void&#160;</td><td class="memItemRight" valign="bottom"><b>SelectParents</b> (<a class="el" href="classpanga_1_1Population.html">Population</a> *population, std::pair&lt; <a class="el" href="classpanga_1_1Individual.html">Individual</a> *, <a class="el" href="classpanga_1_1Individual.html">Individual</a> *&gt; *parents)</td></tr>
<tr class="separator:a1042a5efa8486919bba28548534069da"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad647a9d4cd07a6c26920d607344aa607"><td class="memItemLeft" align="right" valign="top"><a id="ad647a9d4cd07a6c26920d607344aa607"></a>
<a class="el" href="classpanga_1_1Individual.html">Individual</a> *&#160;</td><td class="memItemRight" valign="bottom"><b>SelectOne</b> (<a class="el" href="classpanga_1_1Population.html">Population</a> *population)</td></tr>
<tr class="separator:ad647a9d4cd07a6c26920d607344aa607"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pro-attribs"></a>
Protected Attributes</h2></td></tr>
<tr class="memitem:aa789561b8ec98c18d31580b7c0f2f57e"><td class="memItemLeft" align="right" valign="top"><a id="aa789561b8ec98c18d31580b7c0f2f57e"></a>
const <a class="el" href="classpanga_1_1Genome.html">Genome</a> *&#160;</td><td class="memItemRight" valign="bottom"><b>_genome</b></td></tr>
<tr class="separator:aa789561b8ec98c18d31580b7c0f2f57e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a032daf0d735a8f964c860edf684f1eec"><td class="memItemLeft" align="right" valign="top"><a id="a032daf0d735a8f964c860edf684f1eec"></a>
<a class="el" href="classpanga_1_1Population.html">Population</a>&#160;</td><td class="memItemRight" valign="bottom"><b>_population</b></td></tr>
<tr class="separator:a032daf0d735a8f964c860edf684f1eec"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a1a7fb37d227f8204c2a6a59339d4edf3"><td class="memItemLeft" align="right" valign="top"><a id="a1a7fb37d227f8204c2a6a59339d4edf3"></a>
<a class="el" href="classpanga_1_1Population.html">Population</a>&#160;</td><td class="memItemRight" valign="bottom"><b>_lastGenerationPopulation</b></td></tr>
<tr class="separator:a1a7fb37d227f8204c2a6a59339d4edf3"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a7d59cb63bb71fed40ce09e25ac19d39e"><td class="memItemLeft" align="right" valign="top"><a id="a7d59cb63bb71fed40ce09e25ac19d39e"></a>
<a class="el" href="classpanga_1_1GeneticAlgorithm.html#a84ecf735778d44eceefdc7c9f5a4f21b">MutationRateSchedule</a>&#160;</td><td class="memItemRight" valign="bottom"><b>_mutationRateSchedule</b></td></tr>
<tr class="separator:a7d59cb63bb71fed40ce09e25ac19d39e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a5a3734f127d5de3ba9cfa5c610787bf0"><td class="memItemLeft" align="right" valign="top"><a id="a5a3734f127d5de3ba9cfa5c610787bf0"></a>
size_t&#160;</td><td class="memItemRight" valign="bottom"><b>_populationSize</b></td></tr>
<tr class="separator:a5a3734f127d5de3ba9cfa5c610787bf0"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a7903bacbcade1c9edcd460408507f148"><td class="memItemLeft" align="right" valign="top"><a id="a7903bacbcade1c9edcd460408507f148"></a>
size_t&#160;</td><td class="memItemRight" valign="bottom"><b>_totalGenerations</b></td></tr>
<tr class="separator:a7903bacbcade1c9edcd460408507f148"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a69bf31205e29e3c9c656570876d695f8"><td class="memItemLeft" align="right" valign="top"><a id="a69bf31205e29e3c9c656570876d695f8"></a>
size_t&#160;</td><td class="memItemRight" valign="bottom"><b>_currentGeneration</b></td></tr>
<tr class="separator:a69bf31205e29e3c9c656570876d695f8"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af794a4bcd0d074b87b7c1c07f632a259"><td class="memItemLeft" align="right" valign="top"><a id="af794a4bcd0d074b87b7c1c07f632a259"></a>
size_t&#160;</td><td class="memItemRight" valign="bottom"><b>_eliteCount</b></td></tr>
<tr class="separator:af794a4bcd0d074b87b7c1c07f632a259"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:abd1920c348db00f3120dfd8ea001f43b"><td class="memItemLeft" align="right" valign="top"><a id="abd1920c348db00f3120dfd8ea001f43b"></a>
size_t&#160;</td><td class="memItemRight" valign="bottom"><b>_mutatedEliteCount</b></td></tr>
<tr class="separator:abd1920c348db00f3120dfd8ea001f43b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa7f1008612d502f24c959ff183f4e38c"><td class="memItemLeft" align="right" valign="top"><a id="aa7f1008612d502f24c959ff183f4e38c"></a>
double&#160;</td><td class="memItemRight" valign="bottom"><b>_mutationRate</b></td></tr>
<tr class="separator:aa7f1008612d502f24c959ff183f4e38c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a041c5d94896e2a853f89646fdab2cd3d"><td class="memItemLeft" align="right" valign="top"><a id="a041c5d94896e2a853f89646fdab2cd3d"></a>
double&#160;</td><td class="memItemRight" valign="bottom"><b>_crossoverRate</b></td></tr>
<tr class="separator:a041c5d94896e2a853f89646fdab2cd3d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a1ca456e14493a719f750ace62a76731d"><td class="memItemLeft" align="right" valign="top"><a id="a1ca456e14493a719f750ace62a76731d"></a>
double&#160;</td><td class="memItemRight" valign="bottom"><b>_mutatedEliteMutationRate</b></td></tr>
<tr class="separator:a1ca456e14493a719f750ace62a76731d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae0575ff65aab25ac05430438493d924c"><td class="memItemLeft" align="right" valign="top"><a id="ae0575ff65aab25ac05430438493d924c"></a>
<a class="el" href="classpanga_1_1GeneticAlgorithm.html#a121354609fbc300abfb96556935f4b35">CrossoverType</a>&#160;</td><td class="memItemRight" valign="bottom"><b>_crossoverType</b></td></tr>
<tr class="separator:ae0575ff65aab25ac05430438493d924c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a4659aefce5134e09828e301395bdd0c8"><td class="memItemLeft" align="right" valign="top"><a id="a4659aefce5134e09828e301395bdd0c8"></a>
<a class="el" href="classpanga_1_1GeneticAlgorithm.html#a175a0790ea362ea9794fa4daa222d818">MutatorType</a>&#160;</td><td class="memItemRight" valign="bottom"><b>_mutatorType</b></td></tr>
<tr class="separator:a4659aefce5134e09828e301395bdd0c8"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a1514b278b8a0511a0cfed3184ebec8be"><td class="memItemLeft" align="right" valign="top"><a id="a1514b278b8a0511a0cfed3184ebec8be"></a>
<a class="el" href="classpanga_1_1GeneticAlgorithm.html#a3653afdb62876d5d18254dba4de27a6e">SelectorType</a>&#160;</td><td class="memItemRight" valign="bottom"><b>_selectorType</b></td></tr>
<tr class="separator:a1514b278b8a0511a0cfed3184ebec8be"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac9ddaca2035967b18562a0ccd5ec058d"><td class="memItemLeft" align="right" valign="top"><a id="ac9ddaca2035967b18562a0ccd5ec058d"></a>
size_t&#160;</td><td class="memItemRight" valign="bottom"><b>_tournamentSize</b></td></tr>
<tr class="separator:ac9ddaca2035967b18562a0ccd5ec058d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a83aa25361790ef3293244807ab96f41b"><td class="memItemLeft" align="right" valign="top"><a id="a83aa25361790ef3293244807ab96f41b"></a>
size_t&#160;</td><td class="memItemRight" valign="bottom"><b>_kPointCrossoverPointCount</b></td></tr>
<tr class="separator:a83aa25361790ef3293244807ab96f41b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab47eb2d8d582ef74b40467790df8acdb"><td class="memItemLeft" align="right" valign="top"><a id="ab47eb2d8d582ef74b40467790df8acdb"></a>
double&#160;</td><td class="memItemRight" valign="bottom"><b>_selfAdaptiveMutationDiversityFloor</b></td></tr>
<tr class="separator:ab47eb2d8d582ef74b40467790df8acdb"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a67ad20794f307a59a6b718f80d300c4f"><td class="memItemLeft" align="right" valign="top"><a id="a67ad20794f307a59a6b718f80d300c4f"></a>
double&#160;</td><td class="memItemRight" valign="bottom"><b>_selfAdaptiveMutationAggressiveRate</b></td></tr>
<tr class="separator:a67ad20794f307a59a6b718f80d300c4f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a1cbfc2bec5b743d78e7871f901ece6ef"><td class="memItemLeft" align="right" valign="top"><a id="a1cbfc2bec5b743d78e7871f901ece6ef"></a>
size_t&#160;</td><td class="memItemRight" valign="bottom"><b>_proportionalMutationBitCount</b></td></tr>
<tr class="separator:a1cbfc2bec5b743d78e7871f901ece6ef"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a2bb1265c949306f101ad4af8f7ea4f5d"><td class="memItemLeft" align="right" valign="top"><a id="a2bb1265c949306f101ad4af8f7ea4f5d"></a>
bool&#160;</td><td class="memItemRight" valign="bottom"><b>_crossoverIgnoreGeneBoundaries</b></td></tr>
<tr class="separator:a2bb1265c949306f101ad4af8f7ea4f5d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a64494b6e4b42c0fe645275e21ea63d31"><td class="memItemLeft" align="right" valign="top"><a id="a64494b6e4b42c0fe645275e21ea63d31"></a>
bool&#160;</td><td class="memItemRight" valign="bottom"><b>_allowSameParentCouples</b></td></tr>
<tr class="separator:a64494b6e4b42c0fe645275e21ea63d31"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a9b2d95860716ebd2efd8db41c3543211"><td class="memItemLeft" align="right" valign="top"><a id="a9b2d95860716ebd2efd8db41c3543211"></a>
void *&#160;</td><td class="memItemRight" valign="bottom"><b>_userData</b></td></tr>
<tr class="separator:a9b2d95860716ebd2efd8db41c3543211"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:acfe478da7d72eba50a6c45448c04106a"><td class="memItemLeft" align="right" valign="top"><a id="acfe478da7d72eba50a6c45448c04106a"></a>
FitnessFunction&#160;</td><td class="memItemRight" valign="bottom"><b>_fitnessFunction</b></td></tr>
<tr class="separator:acfe478da7d72eba50a6c45448c04106a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aacb785fc25e8a1b33e4739860c929388"><td class="memItemLeft" align="right" valign="top"><a id="aacb785fc25e8a1b33e4739860c929388"></a>
<a class="el" href="classpanga_1_1RandomWrapper.html">RandomWrapper</a>&#160;</td><td class="memItemRight" valign="bottom"><b>_randomWrapper</b></td></tr>
<tr class="separator:aacb785fc25e8a1b33e4739860c929388"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<h2 class="groupheader">Member Enumeration Documentation</h2>
<a id="a121354609fbc300abfb96556935f4b35"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a121354609fbc300abfb96556935f4b35">&#9670;&nbsp;</a></span>CrossoverType</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">enum <a class="el" href="classpanga_1_1GeneticAlgorithm.html#a121354609fbc300abfb96556935f4b35">panga::GeneticAlgorithm::CrossoverType</a> : uint8_t</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">strong</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<p>Type of crossover we will perform on selected parents in order to produce offspring for the next generation.<br />
 Actual crossover is performed by the <a class="el" href="classpanga_1_1Chromosome.html">Chromosome</a> itself. </p><dl class="section see"><dt>See also</dt><dd>Crossover </dd>
<dd>
<a class="el" href="classpanga_1_1GeneticAlgorithm.html#a070d1761aaa17dd6554cc23a0dc34f97">SetCrossoverType</a> </dd>
<dd>
GetCrossoverType </dd>
<dd>
<a class="el" href="classpanga_1_1Chromosome.html">Chromosome</a> </dd></dl>
<table class="fieldtable">
<tr><th colspan="2">Enumerator</th></tr><tr><td class="fieldname"><a id="a121354609fbc300abfb96556935f4b35af19d8dfb9f04573bad667420b8cb2270"></a>OnePoint&#160;</td><td class="fielddoc"><p>Perform one-point crossover.<br />
 The parent chromosomes will be split into two pieces (based on one cut point) with the offspring receiving the left part of one parent and the right part of the other parent.<br />
 The point at which to cut the parent chromosomes is determined randomly. </p><dl class="section see"><dt>See also</dt><dd>Chromosome::KPointCrossover </dd></dl>
</td></tr>
<tr><td class="fieldname"><a id="a121354609fbc300abfb96556935f4b35a331d16a2885c7f82ba949f286f48143d"></a>TwoPoint&#160;</td><td class="fielddoc"><p>Perform two-point crossover.<br />
 Similar to one-point crossover, but the parent chromosomes will be split into three pieces (based on two cut points) with the offspring receiving the first and last part from one parent and the middle part from the other parent.<br />
 The points at which the parent chromosomes is cut will be determined randomly. </p><dl class="section see"><dt>See also</dt><dd>Chromosome::KPointCrossover </dd></dl>
</td></tr>
<tr><td class="fieldname"><a id="a121354609fbc300abfb96556935f4b35a0d398d8c2d3f38c8884ae132c394b784"></a>KPoint&#160;</td><td class="fielddoc"><p>Perform k-point crossover.<br />
 Similar to two-point crossover, but the parent chromosomes will be split into k+1 pieces (based on k cut points) with the offspring receiving pieces from both parents alternating.<br />
 The points at which the parent chromosomes is cut will be determined randomly.<br />
 k is determined by _kPointCrossoverPointCount. </p><dl class="section see"><dt>See also</dt><dd>_kPointCrossoverPointCount </dd>
<dd>
<a class="el" href="classpanga_1_1GeneticAlgorithm.html#a72a9de6fcb54fc424a4fa5d66f388907">SetKPointCrossoverPointCount</a> </dd>
<dd>
GetKPointCrossoverPointCount </dd>
<dd>
Chromosome::KPointCrossover </dd></dl>
</td></tr>
<tr><td class="fieldname"><a id="a121354609fbc300abfb96556935f4b35af19516d11f2946f894070e92fcb56b6d"></a>Uniform&#160;</td><td class="fielddoc"><p>Perform uniform crossover.<br />
 Each bit in the chromosome has an equal chance of being copied from either parent. </p><dl class="section see"><dt>See also</dt><dd>Chromosome::UniformCrossover </dd></dl>
</td></tr>
</table>

</div>
</div>
<a id="a84ecf735778d44eceefdc7c9f5a4f21b"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a84ecf735778d44eceefdc7c9f5a4f21b">&#9670;&nbsp;</a></span>MutationRateSchedule</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">enum <a class="el" href="classpanga_1_1GeneticAlgorithm.html#a84ecf735778d44eceefdc7c9f5a4f21b">panga::GeneticAlgorithm::MutationRateSchedule</a> : uint8_t</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">strong</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<p>Method by which the mutation rate is adjusted over the course of running the genetic algorithm. </p><dl class="section see"><dt>See also</dt><dd><a class="el" href="classpanga_1_1GeneticAlgorithm.html#a3e3cb4264ee3ae4a892dbd6a2bf43276">SetMutationRateSchedule</a> </dd>
<dd>
GetMutationRateSchedule </dd>
<dd>
GetCurrentMutationRate </dd>
<dd>
<a class="el" href="classpanga_1_1GeneticAlgorithm.html#a4b0b1a82ef3d8e6881927ca96b27947c">SetMutationRate</a> </dd>
<dd>
GetMutationRate </dd></dl>
<table class="fieldtable">
<tr><th colspan="2">Enumerator</th></tr><tr><td class="fieldname"><a id="a84ecf735778d44eceefdc7c9f5a4f21bacb17869fe51048b5a5c4c6106551a255"></a>Constant&#160;</td><td class="fielddoc"><p>The mutation rate is constant and does not vary from generation to generation.<br />
 The rate will always be _mutationRate. </p><dl class="section see"><dt>See also</dt><dd>_mutationRate </dd>
<dd>
<a class="el" href="classpanga_1_1GeneticAlgorithm.html#a4b0b1a82ef3d8e6881927ca96b27947c">SetMutationRate</a> </dd>
<dd>
GetMutationRate </dd></dl>
</td></tr>
<tr><td class="fieldname"><a id="a84ecf735778d44eceefdc7c9f5a4f21bac172d0b7b7d4c61a931d50369d37b11f"></a>Deterministic&#160;</td><td class="fielddoc"><p>The mutation rate varies by generation.<br />
 Early generations introduce a lot of mutation which slowly tapers off to 0 by the last generation. (last generation is controlled by _totalGenerations)<br />
 Strictly, when the current generation has passed _totalGenerations, we will set the current mutation rate to _mutationRate instead of 0 because we do not want to stall the evolution.<br />
 The computed mutation rate has nothing to do with _mutationRate unless we have run-past _totalGenerations.<br />
 This mutation schedule does a good job of using the mutation operator to search the solution-space of the chromosome. </p><dl class="section see"><dt>See also</dt><dd>_totalGenerations </dd>
<dd>
_mutationRate </dd>
<dd>
<a class="el" href="classpanga_1_1GeneticAlgorithm.html#a4b0b1a82ef3d8e6881927ca96b27947c">SetMutationRate</a> </dd>
<dd>
GetMutationRate </dd>
<dd>
<a class="el" href="classpanga_1_1GeneticAlgorithm.html#add6b138984b9541983cad01774a2b3c9">SetTotalGenerations</a> </dd>
<dd>
GetTotalGenerations </dd></dl>
</td></tr>
<tr><td class="fieldname"><a id="a84ecf735778d44eceefdc7c9f5a4f21baf5f333dcdae61f0aa41cd7b8c82222aa"></a>SelfAdaptive&#160;</td><td class="fielddoc"><p>Attempt to increase mutation when the population converges too much. Uses the population diversity score as a metric.<br />
 When this diveristy falls below a floor value (_selfAdaptiveMutationDiversityFloor), we will introduce more aggressive mutation in order to increase diversity and keep the evolution from converging.<br />
 The more aggressive mutation rate is configurable via _selfAdaptiveMutationAggressiveRate.<br />
 When diversity is not below the floor value, we use a constant mutation rate controlled by _mutationRate. </p><dl class="section see"><dt>See also</dt><dd>_mutationRate </dd>
<dd>
<a class="el" href="classpanga_1_1GeneticAlgorithm.html#a4b0b1a82ef3d8e6881927ca96b27947c">SetMutationRate</a> </dd>
<dd>
GetMutationRate </dd>
<dd>
_selfAdaptiveMutationDiversityFloor </dd>
<dd>
<a class="el" href="classpanga_1_1GeneticAlgorithm.html#aca07c13c4b683605270cb2f0ea24a28c">SetSelfAdaptiveMutationDiversityFloor</a> </dd>
<dd>
GetSelfAdaptiveMutationDiversityFloor </dd>
<dd>
_selfAdaptiveMutationAggressiveRate </dd>
<dd>
<a class="el" href="classpanga_1_1GeneticAlgorithm.html#a07a9e112cb63a3a6ddc101a3a31e0411">SetSelfAdaptiveMutationAggressiveRate</a> </dd>
<dd>
GetSelfAdaptiveMutationAggressiveRate </dd></dl>
</td></tr>
<tr><td class="fieldname"><a id="a84ecf735778d44eceefdc7c9f5a4f21baac13b33ebc8ea372dfb07e5ed4d9655f"></a>Proportional&#160;</td><td class="fielddoc"><p>Calculates the mutation rate required to flip a number of bits in the chromosome.<br />
 The number of bits is controlled by _proportionalMutationBitCount. </p><dl class="section see"><dt>See also</dt><dd>_proportionalMutationBitCount </dd>
<dd>
<a class="el" href="classpanga_1_1GeneticAlgorithm.html#a536b029dd25bdd78a17fcbefd21f8b11">SetProportionalMutationBitCount</a> </dd>
<dd>
GetProportionalMutationBitCount </dd></dl>
</td></tr>
</table>

</div>
</div>
<a id="a175a0790ea362ea9794fa4daa222d818"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a175a0790ea362ea9794fa4daa222d818">&#9670;&nbsp;</a></span>MutatorType</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">enum <a class="el" href="classpanga_1_1GeneticAlgorithm.html#a175a0790ea362ea9794fa4daa222d818">panga::GeneticAlgorithm::MutatorType</a> : uint8_t</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">strong</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<p>The type of mutation we will perform on new individuals when they are added to the next generation population. </p><dl class="section see"><dt>See also</dt><dd>Mutate </dd>
<dd>
<a class="el" href="classpanga_1_1GeneticAlgorithm.html#ae43cef679831a1917d13795401387250">SetMutatorType</a> </dd>
<dd>
GetMutatorType </dd></dl>
<table class="fieldtable">
<tr><th colspan="2">Enumerator</th></tr><tr><td class="fieldname"><a id="a175a0790ea362ea9794fa4daa222d818a9ffbd422925a6839ee820ddbc59278c5"></a>Flip&#160;</td><td class="fielddoc"><p>Perform the flip mutator.<br />
 This randomly flips bits in the <a class="el" href="classpanga_1_1Individual.html">Individual</a> based on the mutation rate provided.<br />
 Every bit in the chromosome has a mutation rate chance of flipping. </p><dl class="section see"><dt>See also</dt><dd>Chromosome::FlipMutator </dd></dl>
</td></tr>
</table>

</div>
</div>
<a id="a3653afdb62876d5d18254dba4de27a6e"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a3653afdb62876d5d18254dba4de27a6e">&#9670;&nbsp;</a></span>SelectorType</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">enum <a class="el" href="classpanga_1_1GeneticAlgorithm.html#a3653afdb62876d5d18254dba4de27a6e">panga::GeneticAlgorithm::SelectorType</a> : uint8_t</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">strong</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<p>Algorithm we want to use when we select individuals from the current population to become parents of an offspring in the next generation population.<br />
 Actual selection is done by the <a class="el" href="classpanga_1_1Population.html">Population</a> itself. </p><dl class="section see"><dt>See also</dt><dd><a class="el" href="classpanga_1_1GeneticAlgorithm.html#ab0764a6c0e0db8cd6a82d109a9789e9a">SetSelectorType</a> </dd>
<dd>
GetSelectorType </dd>
<dd>
SelectParents </dd>
<dd>
SelectOne </dd>
<dd>
<a class="el" href="classpanga_1_1Population.html">Population</a> </dd></dl>
<table class="fieldtable">
<tr><th colspan="2">Enumerator</th></tr><tr><td class="fieldname"><a id="a3653afdb62876d5d18254dba4de27a6ea021da1b20f73dc252361a54d80497ef3"></a>Rank&#160;</td><td class="fielddoc"><p>Rank selector.<br />
 Always returns the fittest individual from a population.<br />
 When _allowSameParentCouples is false, returns the top two individuals when we select a couple from a population. </p><dl class="section see"><dt>See also</dt><dd>Population::RankSelect </dd></dl>
</td></tr>
<tr><td class="fieldname"><a id="a3653afdb62876d5d18254dba4de27a6eaf19516d11f2946f894070e92fcb56b6d"></a>Uniform&#160;</td><td class="fielddoc"><p>Uniform selector.<br />
 Always returns a random individual from a population. </p><dl class="section see"><dt>See also</dt><dd>Population::UniformSelect </dd></dl>
</td></tr>
<tr><td class="fieldname"><a id="a3653afdb62876d5d18254dba4de27a6eaf1c787f267b4d6d03f1202063099a24a"></a>RouletteWheel&#160;</td><td class="fielddoc"><p>Roulette wheel selector.<br />
 Selects an individual with higher fitness more often but has a chance to select any individual in a population.<br />
 This can be visualized by imagining a roulette wheel where each individual is assigned one slice of the wheel. The size of each slice is determined by the fitness of the individual where more fit individuals are assigned a larger slice. Then the wheel is spun and whichever slice is chosen will cause the associated individual to be selected. </p><dl class="section see"><dt>See also</dt><dd>InitializeSelector </dd>
<dd>
Population::RouletteWheelSelect </dd></dl>
</td></tr>
<tr><td class="fieldname"><a id="a3653afdb62876d5d18254dba4de27a6ea4f4dc29df9d29cbec61a725ff3ce6e72"></a>Tournament&#160;</td><td class="fielddoc"><p>Tournament selector.<br />
 Chooses n individuals randomly from the population and selects the one with the highest fitness.<br />
 This tends to select individuals with higher fitness but the tournament size has an effect.<br />
 If n is 1, tournament selector is identical to uniform selector.<br />
 If n is the population size, tournament selector is identical to rank selector.<br />
 You can control n with the _tournamentSize property. </p><dl class="section see"><dt>See also</dt><dd>_tournamentSize </dd>
<dd>
<a class="el" href="classpanga_1_1GeneticAlgorithm.html#ad6deca0de2a42a646a5d2b48f83a12b6">SetTournamentSize</a> </dd>
<dd>
GetTournamentSize </dd>
<dd>
Population::TournamentSelect </dd></dl>
</td></tr>
</table>

</div>
</div>
<h2 class="groupheader">Member Function Documentation</h2>
<a id="a71413b202033cabb4b2c12e2a9e0e04a"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a71413b202033cabb4b2c12e2a9e0e04a">&#9670;&nbsp;</a></span>DeletePopulation()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">void GeneticAlgorithm::DeletePopulation </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="classpanga_1_1Population.html">Population</a> *&#160;</td>
          <td class="paramname"><em>population</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<p>Delete each <a class="el" href="classpanga_1_1Individual.html">Individual</a> in the population and reset the population size to 0. </p>

</div>
</div>
<a id="a36729bf431bb5887ba64275754e09723"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a36729bf431bb5887ba64275754e09723">&#9670;&nbsp;</a></span>Evaluate()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">void GeneticAlgorithm::Evaluate </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="classpanga_1_1Population.html">Population</a> *&#160;</td>
          <td class="paramname"><em>population</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<p>Score each <a class="el" href="classpanga_1_1Individual.html">Individual</a> in the population and then sort the population in terms of decreasing fitness. </p>

</div>
</div>
<a id="a78e68efaff7967fb5937b680f3966353"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a78e68efaff7967fb5937b680f3966353">&#9670;&nbsp;</a></span>GetAverageScore()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">double GeneticAlgorithm::GetAverageScore </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Get the average score between individuals in the current population. </p>

</div>
</div>
<a id="aaf9685a149c8c2102e9ccc54c43d09b4"></a>
<h2 class="memtitle"><span class="permalink"><a href="#aaf9685a149c8c2102e9ccc54c43d09b4">&#9670;&nbsp;</a></span>GetBestIndividual()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classpanga_1_1Individual.html">Individual</a> * GeneticAlgorithm::GetBestIndividual </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Return the best <a class="el" href="classpanga_1_1Individual.html">Individual</a> from the current population. </p>

</div>
</div>
<a id="a40cbc10412d0bcc71f1b9bc2a6d1ebaf"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a40cbc10412d0bcc71f1b9bc2a6d1ebaf">&#9670;&nbsp;</a></span>GetCurrentGeneration()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">size_t GeneticAlgorithm::GetCurrentGeneration </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Return the current generation.<br />
 The generation increases after every step. </p>

</div>
</div>
<a id="a0791b5149f23eb2c397592d8db5ebe19"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a0791b5149f23eb2c397592d8db5ebe19">&#9670;&nbsp;</a></span>GetIndividual()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classpanga_1_1Individual.html">Individual</a> * GeneticAlgorithm::GetIndividual </td>
          <td>(</td>
          <td class="paramtype">size_t&#160;</td>
          <td class="paramname"><em>index</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Return the <a class="el" href="classpanga_1_1Individual.html">Individual</a> at index position from the current population. </p>

</div>
</div>
<a id="af3c2d5f24c373871614aa2ff80c07289"></a>
<h2 class="memtitle"><span class="permalink"><a href="#af3c2d5f24c373871614aa2ff80c07289">&#9670;&nbsp;</a></span>GetMinimumScore()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">double GeneticAlgorithm::GetMinimumScore </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Get the minimum score among individuals in the current population. </p>

</div>
</div>
<a id="a573c741a7c511c58ca1ef0fad1935ee1"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a573c741a7c511c58ca1ef0fad1935ee1">&#9670;&nbsp;</a></span>GetPopulationDiversity()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">double GeneticAlgorithm::GetPopulationDiversity </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Get the population diversity among individuals in the current population. <br />
 The population diversity is a metric used to indicate how much the genetic material backing the individuals varies. A high diversity value means the individuals have very different genetic components. A value of zero means the individuals are identical. </p>

</div>
</div>
<a id="adcc513420e5c5808442c1d9c84932a22"></a>
<h2 class="memtitle"><span class="permalink"><a href="#adcc513420e5c5808442c1d9c84932a22">&#9670;&nbsp;</a></span>GetScoreStandardDeviation()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">double GeneticAlgorithm::GetScoreStandardDeviation </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Get the standard deviation of scores between individuals in the current population. </p>

</div>
</div>
<a id="af5317922e4f09dbe17a2f59b9dd615c2"></a>
<h2 class="memtitle"><span class="permalink"><a href="#af5317922e4f09dbe17a2f59b9dd615c2">&#9670;&nbsp;</a></span>Initialize()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void GeneticAlgorithm::Initialize </td>
          <td>(</td>
          <td class="paramtype">const std::vector&lt; const <a class="el" href="classpanga_1_1BitVector.html">BitVector</a> *&gt; *&#160;</td>
          <td class="paramname"><em>initialPopulation</em> = <code>nullptr</code></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Initialize the state of the <a class="el" href="classpanga_1_1GeneticAlgorithm.html">GeneticAlgorithm</a>.<br />
 If initialPopulation is passed, use these bits as the chromosomes for our new population. Otherwise, create a population filled with random individuals.<br />
 </p><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">initialPopulation</td><td>We will construct new population members and interpret these values as binary chromosome data. Pass nullptr to initialize the population randomly. </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a id="aac0e857b86993a430331bbff5299b231"></a>
<h2 class="memtitle"><span class="permalink"><a href="#aac0e857b86993a430331bbff5299b231">&#9670;&nbsp;</a></span>Run()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void GeneticAlgorithm::Run </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Run the <a class="el" href="classpanga_1_1GeneticAlgorithm.html">GeneticAlgorithm</a> totalGenerations steps. </p>

</div>
</div>
<a id="a08b90a76a83015c56a37b9a2cff28129"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a08b90a76a83015c56a37b9a2cff28129">&#9670;&nbsp;</a></span>Score()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">void GeneticAlgorithm::Score </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="classpanga_1_1Individual.html">Individual</a> *&#160;</td>
          <td class="paramname"><em>individual</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">protected</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<p>Use the fitness function to score one <a class="el" href="classpanga_1_1Individual.html">Individual</a>. </p>

</div>
</div>
<a id="a56014a62f4853e76701abbdc3db0cb18"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a56014a62f4853e76701abbdc3db0cb18">&#9670;&nbsp;</a></span>SetAllowSameParentCouples()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void GeneticAlgorithm::SetAllowSameParentCouples </td>
          <td>(</td>
          <td class="paramtype">bool&#160;</td>
          <td class="paramname"><em>allowSameParentCouples</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>When we choose parents to use for crossover, should we allow the same <a class="el" href="classpanga_1_1Individual.html">Individual</a> to be used as both parents?<br />
 This could be useful because crossover returns the same parent if both parents are identical.<br />
 If this flag is false, we will not allow the same <a class="el" href="classpanga_1_1Individual.html">Individual</a> to be selected as both parents when we select a couple for crossover.<br />
 If this flag is true, it is possible that both parents in the couple selected for crossover will be the same <a class="el" href="classpanga_1_1Individual.html">Individual</a>. This flag is true by default. </p>

</div>
</div>
<a id="af5db5af0bac1cd2923cfc8dab414ba59"></a>
<h2 class="memtitle"><span class="permalink"><a href="#af5db5af0bac1cd2923cfc8dab414ba59">&#9670;&nbsp;</a></span>SetCrossoverIgnoreGeneBoundaries()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void GeneticAlgorithm::SetCrossoverIgnoreGeneBoundaries </td>
          <td>(</td>
          <td class="paramtype">bool&#160;</td>
          <td class="paramname"><em>crossoverIgnoreGeneBoundaries</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>When we perform crossover, should we respect gene boundaries such that all of the bits making up a gene are copied from one parent?<br />
 This could be useful if the bits underlying the gene don't make as much sense when they are split up.<br />
 If this flag is false, we will respect the boundary of each gene and not split a gene up during crossover.<br />
 If this flag is true, we will ignore gene boundaries during crossover and treat the chromosome as a big chunk of binary data which will be split according to the crossover operator chosen.<br />
 In general, this flag should be left at the default of true. </p>

</div>
</div>
<a id="a4a3351748f07e44a99e86a6ac0f85f27"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a4a3351748f07e44a99e86a6ac0f85f27">&#9670;&nbsp;</a></span>SetCrossoverRate()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void GeneticAlgorithm::SetCrossoverRate </td>
          <td>(</td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>crossoverRate</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Set the crossover rate.<br />
 During each step, construct new individuals for the next generation.<br />
 We will select two Individuals and perform crossover to produce the offspring in the next generation with a crossover rate chance.<br />
 Instead of crossover, we will copy one <a class="el" href="classpanga_1_1Individual.html">Individual</a> from the old population over into the new one with (1 - crossover rate) chance. </p>

</div>
</div>
<a id="a070d1761aaa17dd6554cc23a0dc34f97"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a070d1761aaa17dd6554cc23a0dc34f97">&#9670;&nbsp;</a></span>SetCrossoverType()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void GeneticAlgorithm::SetCrossoverType </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="classpanga_1_1GeneticAlgorithm.html#a121354609fbc300abfb96556935f4b35">CrossoverType</a>&#160;</td>
          <td class="paramname"><em>crossoverType</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Set the crossover type.<br />
 We will use this type to crossover two parent individuals in order to produce offspring for the next generation. </p><dl class="section see"><dt>See also</dt><dd><a class="el" href="classpanga_1_1GeneticAlgorithm.html#a121354609fbc300abfb96556935f4b35">CrossoverType</a> </dd></dl>

</div>
</div>
<a id="a9df556f467a9884b449708e03316b302"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a9df556f467a9884b449708e03316b302">&#9670;&nbsp;</a></span>SetEliteCount()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void GeneticAlgorithm::SetEliteCount </td>
          <td>(</td>
          <td class="paramtype">size_t&#160;</td>
          <td class="paramname"><em>eliteCount</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>When constructing the next generation, always copy some of the best Individuals from the current generation. These elite Individuals are copied as they are and are not crossed-over or mutated. </p><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">eliteCount</td><td>Number of elites. </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a id="a92d0aa5d3796aca95e84897323328667"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a92d0aa5d3796aca95e84897323328667">&#9670;&nbsp;</a></span>SetFitnessFunction()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void GeneticAlgorithm::SetFitnessFunction </td>
          <td>(</td>
          <td class="paramtype">FitnessFunction&#160;</td>
          <td class="paramname"><em>fitnessFunction</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Set the fitness function used to evaluate the fitness of each <a class="el" href="classpanga_1_1Individual.html">Individual</a>. <br />
 The fitness function takes a pointer to an <a class="el" href="classpanga_1_1Individual.html">Individual</a> and a void* which is the user data previously set via <a class="el" href="classpanga_1_1GeneticAlgorithm.html#af2f1560069922de9478b57e6d6960b93">GeneticAlgorithm::SetUserData</a>. </p>

</div>
</div>
<a id="aeaf71c63931a61074b12d73f995e8699"></a>
<h2 class="memtitle"><span class="permalink"><a href="#aeaf71c63931a61074b12d73f995e8699">&#9670;&nbsp;</a></span>SetGenome()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void GeneticAlgorithm::SetGenome </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="classpanga_1_1Genome.html">Genome</a> *&#160;</td>
          <td class="paramname"><em>genome</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Set the genome we will use to construct Individuals which make up the populations of each generation. </p>

</div>
</div>
<a id="a72a9de6fcb54fc424a4fa5d66f388907"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a72a9de6fcb54fc424a4fa5d66f388907">&#9670;&nbsp;</a></span>SetKPointCrossoverPointCount()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void GeneticAlgorithm::SetKPointCrossoverPointCount </td>
          <td>(</td>
          <td class="paramtype">size_t&#160;</td>
          <td class="paramname"><em>kPointCrossoverPointCount</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Set the number of points we'll use to cut up the parent chromosomes during k-point crossover. </p>

</div>
</div>
<a id="aa1c0bbfc523f2b72fa7af30fc2a95641"></a>
<h2 class="memtitle"><span class="permalink"><a href="#aa1c0bbfc523f2b72fa7af30fc2a95641">&#9670;&nbsp;</a></span>SetMutatedEliteCount()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void GeneticAlgorithm::SetMutatedEliteCount </td>
          <td>(</td>
          <td class="paramtype">size_t&#160;</td>
          <td class="paramname"><em>mutatedEliteCount</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Mutated elites are the same as ordinary elite Individuals except they are subjected to the mutation operator according to mutatedEliteMutationRate. </p><dl class="section see"><dt>See also</dt><dd><a class="el" href="classpanga_1_1GeneticAlgorithm.html#a4a8bd65c3dd58a49792ab370e398236b">SetMutatedEliteMutationRate</a> </dd></dl>

</div>
</div>
<a id="a4a8bd65c3dd58a49792ab370e398236b"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a4a8bd65c3dd58a49792ab370e398236b">&#9670;&nbsp;</a></span>SetMutatedEliteMutationRate()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void GeneticAlgorithm::SetMutatedEliteMutationRate </td>
          <td>(</td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>mutatedEliteMutationRate</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Set the rate at mutated elite individuals will be mutated. </p>

</div>
</div>
<a id="a4b0b1a82ef3d8e6881927ca96b27947c"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a4b0b1a82ef3d8e6881927ca96b27947c">&#9670;&nbsp;</a></span>SetMutationRate()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void GeneticAlgorithm::SetMutationRate </td>
          <td>(</td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>mutationRate</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Set the mutation rate.<br />
 After we perform crossover, mutate each offspring <a class="el" href="classpanga_1_1Individual.html">Individual</a> by this rate according to the mutation operator.<br />
 During execution, the actual mutation rate is calculated based on the mutation rate schedule. </p><dl class="section see"><dt>See also</dt><dd><a class="el" href="classpanga_1_1GeneticAlgorithm.html#a175a0790ea362ea9794fa4daa222d818">MutatorType</a> </dd>
<dd>
<a class="el" href="classpanga_1_1GeneticAlgorithm.html#a84ecf735778d44eceefdc7c9f5a4f21b">MutationRateSchedule</a> </dd>
<dd>
GetCurrentMutationRate </dd></dl>

</div>
</div>
<a id="a3e3cb4264ee3ae4a892dbd6a2bf43276"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a3e3cb4264ee3ae4a892dbd6a2bf43276">&#9670;&nbsp;</a></span>SetMutationRateSchedule()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void GeneticAlgorithm::SetMutationRateSchedule </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="classpanga_1_1GeneticAlgorithm.html#a84ecf735778d44eceefdc7c9f5a4f21b">MutationRateSchedule</a>&#160;</td>
          <td class="paramname"><em>mutationRateSchedule</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Set the mutation rate schedule we will use to determine the mutation rate over the course of many generations. </p><dl class="section see"><dt>See also</dt><dd><a class="el" href="classpanga_1_1GeneticAlgorithm.html#a84ecf735778d44eceefdc7c9f5a4f21b">MutationRateSchedule</a> </dd></dl>

</div>
</div>
<a id="ae43cef679831a1917d13795401387250"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ae43cef679831a1917d13795401387250">&#9670;&nbsp;</a></span>SetMutatorType()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void GeneticAlgorithm::SetMutatorType </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="classpanga_1_1GeneticAlgorithm.html#a175a0790ea362ea9794fa4daa222d818">MutatorType</a>&#160;</td>
          <td class="paramname"><em>mutatorType</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Set the mutator type.<br />
 We will use the mutator to mutate each new offspring before adding it to the next generation population. </p><dl class="section see"><dt>See also</dt><dd><a class="el" href="classpanga_1_1GeneticAlgorithm.html#a175a0790ea362ea9794fa4daa222d818">MutatorType</a> </dd></dl>

</div>
</div>
<a id="ae1f5aad716fae81c066428b6f2069d25"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ae1f5aad716fae81c066428b6f2069d25">&#9670;&nbsp;</a></span>SetPopulationSize()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void GeneticAlgorithm::SetPopulationSize </td>
          <td>(</td>
          <td class="paramtype">size_t&#160;</td>
          <td class="paramname"><em>populationSize</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Each generation, we will construct and evaluate populationSize Individuals. </p>

</div>
</div>
<a id="a536b029dd25bdd78a17fcbefd21f8b11"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a536b029dd25bdd78a17fcbefd21f8b11">&#9670;&nbsp;</a></span>SetProportionalMutationBitCount()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void GeneticAlgorithm::SetProportionalMutationBitCount </td>
          <td>(</td>
          <td class="paramtype">size_t&#160;</td>
          <td class="paramname"><em>proportionalMutationBitCount</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Set the number of bits we should attempt to mutate with each mutation operation. </p>

</div>
</div>
<a id="ab0764a6c0e0db8cd6a82d109a9789e9a"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ab0764a6c0e0db8cd6a82d109a9789e9a">&#9670;&nbsp;</a></span>SetSelectorType()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void GeneticAlgorithm::SetSelectorType </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="classpanga_1_1GeneticAlgorithm.html#a3653afdb62876d5d18254dba4de27a6e">SelectorType</a>&#160;</td>
          <td class="paramname"><em>selectorType</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Set the selector type.<br />
 We will use the selector specified to select two parent individuals on which to perform crossover to produce offspring for the next generation. </p><dl class="section see"><dt>See also</dt><dd><a class="el" href="classpanga_1_1GeneticAlgorithm.html#a3653afdb62876d5d18254dba4de27a6e">SelectorType</a> </dd></dl>

</div>
</div>
<a id="a07a9e112cb63a3a6ddc101a3a31e0411"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a07a9e112cb63a3a6ddc101a3a31e0411">&#9670;&nbsp;</a></span>SetSelfAdaptiveMutationAggressiveRate()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void GeneticAlgorithm::SetSelfAdaptiveMutationAggressiveRate </td>
          <td>(</td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>selfAdaptiveMutationAggressiveRate</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Set the aggressive mutation rate we will switch to when the population diversity falls below the self-adaptive mutation diversity floor. </p>

</div>
</div>
<a id="aca07c13c4b683605270cb2f0ea24a28c"></a>
<h2 class="memtitle"><span class="permalink"><a href="#aca07c13c4b683605270cb2f0ea24a28c">&#9670;&nbsp;</a></span>SetSelfAdaptiveMutationDiversityFloor()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void GeneticAlgorithm::SetSelfAdaptiveMutationDiversityFloor </td>
          <td>(</td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>selfAdaptiveMutationDiversityFloor</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Set the floor value for the population diversity, below which we will more aggressively mutate the population. </p>

</div>
</div>
<a id="add6b138984b9541983cad01774a2b3c9"></a>
<h2 class="memtitle"><span class="permalink"><a href="#add6b138984b9541983cad01774a2b3c9">&#9670;&nbsp;</a></span>SetTotalGenerations()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void GeneticAlgorithm::SetTotalGenerations </td>
          <td>(</td>
          <td class="paramtype">size_t&#160;</td>
          <td class="paramname"><em>totalGenerations</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Set the total number of generations we want to evolve.<br />
 If you call <a class="el" href="classpanga_1_1GeneticAlgorithm.html#aac0e857b86993a430331bbff5299b231">Run()</a>, we will execute this many generations.<br />
 If you manually call <a class="el" href="classpanga_1_1GeneticAlgorithm.html#a86aaa0a83bc03d318d0c6a4d30c9dbeb">Step()</a>, you can run more than total generations. <br />
 This number is used to calculate mutation rate in some mutation rate schedules. </p>

</div>
</div>
<a id="ad6deca0de2a42a646a5d2b48f83a12b6"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ad6deca0de2a42a646a5d2b48f83a12b6">&#9670;&nbsp;</a></span>SetTournamentSize()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void GeneticAlgorithm::SetTournamentSize </td>
          <td>(</td>
          <td class="paramtype">size_t&#160;</td>
          <td class="paramname"><em>tournamentSize</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Set the number of Individuals we should include in the tournament when using tournament selector. </p>

</div>
</div>
<a id="af2f1560069922de9478b57e6d6960b93"></a>
<h2 class="memtitle"><span class="permalink"><a href="#af2f1560069922de9478b57e6d6960b93">&#9670;&nbsp;</a></span>SetUserData()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void GeneticAlgorithm::SetUserData </td>
          <td>(</td>
          <td class="paramtype">void *&#160;</td>
          <td class="paramname"><em>userData</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Set some data which will be passed into the fitness function called to score each <a class="el" href="classpanga_1_1Individual.html">Individual</a>. </p>

</div>
</div>
<a id="a86aaa0a83bc03d318d0c6a4d30c9dbeb"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a86aaa0a83bc03d318d0c6a4d30c9dbeb">&#9670;&nbsp;</a></span>Step()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void GeneticAlgorithm::Step </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Perform one step of the genetic algorithm:<br />
</p><ol type="1">
<li>Score the current population of individuals.<br />
<ul>
<li>This will either be the initial population (if current generation is 0) or the population we constructed during the previous step operation.<br />
</li>
<li>The score operation will leave the population sorted.<br />
</li>
</ul>
</li>
<li>Use the current population to construct the next generation population.<br />
<ul>
<li>Uses a combination of elitism, crossover, and mutation to construct the new population.<br />
</li>
<li>Individuals are chosen from the current population based on their fitness scores.<br />
</li>
<li>The next generation population is not scored, yet. It will be during the next step operation.<br />
</li>
</ul>
</li>
<li>Advance the current generation. </li>
</ol>

</div>
</div>
<hr/>The documentation for this class was generated from the following files:<ul>
<li>src/<a class="el" href="GeneticAlgorithm_8h_source.html">GeneticAlgorithm.h</a></li>
<li>src/GeneticAlgorithm.cc</li>
</ul>
</div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by &#160;<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/>
</a> 1.8.13
</small></address>
</body>
</html>
