<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.8.13"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>panga: panga::Genome Class Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">panga
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.13 -->
<script type="text/javascript">
var searchBox = new SearchBox("searchBox", "search",false,'Search');
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
</script>
<div id="main-nav"></div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div id="nav-path" class="navpath">
  <ul>
<li class="navelem"><b>panga</b></li><li class="navelem"><a class="el" href="classpanga_1_1Genome.html">Genome</a></li>  </ul>
</div>
</div><!-- top -->
<div class="header">
  <div class="summary">
<a href="#pub-methods">Public Member Functions</a> &#124;
<a href="#pro-attribs">Protected Attributes</a> &#124;
<a href="classpanga_1_1Genome-members.html">List of all members</a>  </div>
  <div class="headertitle">
<div class="title">panga::Genome Class Reference</div>  </div>
</div><!--header-->
<div class="contents">

<p><code>#include &lt;<a class="el" href="Genome_8h_source.html">Genome.h</a>&gt;</code></p>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-methods"></a>
Public Member Functions</h2></td></tr>
<tr class="memitem:ab5b50407e016a4bb3f921f94b2ebbfb4"><td class="memTemplParams" colspan="2">template&lt;bool byteAlign = false&gt; </td></tr>
<tr class="memitem:ab5b50407e016a4bb3f921f94b2ebbfb4"><td class="memTemplItemLeft" align="right" valign="top">size_t&#160;</td><td class="memTemplItemRight" valign="bottom"><a class="el" href="classpanga_1_1Genome.html#ab5b50407e016a4bb3f921f94b2ebbfb4">AddGene</a> (size_t geneBitWidth)</td></tr>
<tr class="separator:ab5b50407e016a4bb3f921f94b2ebbfb4"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:adf73367588c8932a4379e9d2d92d02ff"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpanga_1_1Genome.html#adf73367588c8932a4379e9d2d92d02ff">SetBooleanGeneCount</a> (size_t booleanGeneCount)</td></tr>
<tr class="separator:adf73367588c8932a4379e9d2d92d02ff"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a915b97071a216d1ccf01f064af888be8"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpanga_1_1Genome.html#a915b97071a216d1ccf01f064af888be8">AddBooleanGenes</a> (size_t count)</td></tr>
<tr class="separator:a915b97071a216d1ccf01f064af888be8"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a8a2f7f6c10a0e96addd4a71a84d699cf"><td class="memItemLeft" align="right" valign="top">size_t&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpanga_1_1Genome.html#a8a2f7f6c10a0e96addd4a71a84d699cf">GetBooleanGeneCount</a> () const</td></tr>
<tr class="separator:a8a2f7f6c10a0e96addd4a71a84d699cf"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a81ef777300e2c0861c69de1421baab8c"><td class="memItemLeft" align="right" valign="top">size_t&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpanga_1_1Genome.html#a81ef777300e2c0861c69de1421baab8c">GetGeneStartBitIndex</a> (size_t geneIndex) const</td></tr>
<tr class="separator:a81ef777300e2c0861c69de1421baab8c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac9db69ece5360c5eaa94c9cbc8b5d0b6"><td class="memItemLeft" align="right" valign="top">size_t&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpanga_1_1Genome.html#ac9db69ece5360c5eaa94c9cbc8b5d0b6">GetGeneBitWitdh</a> (size_t geneIndex) const</td></tr>
<tr class="separator:ac9db69ece5360c5eaa94c9cbc8b5d0b6"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a56f745762146d7579823b54d4acdbc9f"><td class="memItemLeft" align="right" valign="top">size_t&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpanga_1_1Genome.html#a56f745762146d7579823b54d4acdbc9f">GetFirstBooleanGeneBitIndex</a> () const</td></tr>
<tr class="separator:a56f745762146d7579823b54d4acdbc9f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af5432ac15e82a2c8693affa88375b202"><td class="memItemLeft" align="right" valign="top">size_t&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpanga_1_1Genome.html#af5432ac15e82a2c8693affa88375b202">GetFirstBooleanGeneIndex</a> () const</td></tr>
<tr class="separator:af5432ac15e82a2c8693affa88375b202"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a0440fd8216d128563f22e1d4e1d17303"><td class="memItemLeft" align="right" valign="top">size_t&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpanga_1_1Genome.html#a0440fd8216d128563f22e1d4e1d17303">GetGeneCount</a> () const</td></tr>
<tr class="separator:a0440fd8216d128563f22e1d4e1d17303"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a54c4cb4c89662ddd701716235147ba5e"><td class="memItemLeft" align="right" valign="top">size_t&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpanga_1_1Genome.html#a54c4cb4c89662ddd701716235147ba5e">BitsRequired</a> () const</td></tr>
<tr class="separator:a54c4cb4c89662ddd701716235147ba5e"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pro-attribs"></a>
Protected Attributes</h2></td></tr>
<tr class="memitem:ac909db2e9e05601e0c7715482470aee4"><td class="memItemLeft" align="right" valign="top"><a id="ac909db2e9e05601e0c7715482470aee4"></a>
std::vector&lt; size_t &gt;&#160;</td><td class="memItemRight" valign="bottom"><b>_geneStartBitIndex</b></td></tr>
<tr class="separator:ac909db2e9e05601e0c7715482470aee4"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a673aebff4eb2a2238fc6d86de880a57f"><td class="memItemLeft" align="right" valign="top"><a id="a673aebff4eb2a2238fc6d86de880a57f"></a>
std::vector&lt; size_t &gt;&#160;</td><td class="memItemRight" valign="bottom"><b>_geneBitWidth</b></td></tr>
<tr class="separator:a673aebff4eb2a2238fc6d86de880a57f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac725051d5524fd6d32c5052514779c94"><td class="memItemLeft" align="right" valign="top"><a id="ac725051d5524fd6d32c5052514779c94"></a>
size_t&#160;</td><td class="memItemRight" valign="bottom"><b>_firstBooleanGeneBitIndex</b></td></tr>
<tr class="separator:ac725051d5524fd6d32c5052514779c94"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a59fae85c5a525af3f06194cbc4b973cb"><td class="memItemLeft" align="right" valign="top"><a id="a59fae85c5a525af3f06194cbc4b973cb"></a>
size_t&#160;</td><td class="memItemRight" valign="bottom"><b>_booleanGeneCount</b></td></tr>
<tr class="separator:a59fae85c5a525af3f06194cbc4b973cb"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><p>Provides a representation of the genes which, taken together, represent a genome for members of a species.<br />
 Using the <a class="el" href="classpanga_1_1Genome.html">Genome</a>, we may decode a block of binary data and extract the bits representing each gene from the data. (This is what <a class="el" href="classpanga_1_1Chromosome.html">Chromosome</a> does).<br />
 The <a class="el" href="classpanga_1_1Genome.html">Genome</a> data may be constructed by adding genes one-at-a-time via AddGene along with an optional number of boolean genes.<br />
 The genome is structured into two pieces:<br />
</p><ol type="1">
<li>Genes with a bit width<br />
</li>
<li>Boolean genes which have only one bit value<br />
 For performance reasons, all of the boolean genes are located at the end of the <a class="el" href="classpanga_1_1Genome.html">Genome</a>. It is preferable to add a boolean gene than add a gene with bit width of 1. <dl class="section see"><dt>See also</dt><dd><a class="el" href="classpanga_1_1Genome.html#ab5b50407e016a4bb3f921f94b2ebbfb4">AddGene</a> </dd>
<dd>
<a class="el" href="classpanga_1_1Chromosome.html">Chromosome</a> </dd></dl>
</li>
</ol>
</div><h2 class="groupheader">Member Function Documentation</h2>
<a id="a915b97071a216d1ccf01f064af888be8"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a915b97071a216d1ccf01f064af888be8">&#9670;&nbsp;</a></span>AddBooleanGenes()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Genome::AddBooleanGenes </td>
          <td>(</td>
          <td class="paramtype">size_t&#160;</td>
          <td class="paramname"><em>count</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Add count boolean genes to the boolean gene count for this <a class="el" href="classpanga_1_1Genome.html">Genome</a>.<br />
 Does not return the gene indices because adding more non-boolean genes to the <a class="el" href="classpanga_1_1Genome.html">Genome</a> can change the boolean gene indices. </p><dl class="section see"><dt>See also</dt><dd><a class="el" href="classpanga_1_1Genome.html#af5432ac15e82a2c8693affa88375b202">GetFirstBooleanGeneIndex</a> </dd>
<dd>
<a class="el" href="classpanga_1_1Genome.html#a56f745762146d7579823b54d4acdbc9f">GetFirstBooleanGeneBitIndex</a> </dd></dl>

</div>
</div>
<a id="ab5b50407e016a4bb3f921f94b2ebbfb4"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ab5b50407e016a4bb3f921f94b2ebbfb4">&#9670;&nbsp;</a></span>AddGene()</h2>

<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;bool byteAlign = false&gt; </div>
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">size_t panga::Genome::AddGene </td>
          <td>(</td>
          <td class="paramtype">size_t&#160;</td>
          <td class="paramname"><em>geneBitWidth</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<p>Add a gene to the <a class="el" href="classpanga_1_1Genome.html">Genome</a>.<br />
 The new gene will be added to the end of the <a class="el" href="classpanga_1_1Genome.html">Genome</a>.<br />
 If byteAlign is true, we will shift the starting offset and bit width such that this gene begins and ends on a byte-boundary.<br />
 This can be very useful if you want to access the memory backing this gene later directly (via something like <a class="el" href="classpanga_1_1Chromosome.html#abe6e9d64ff7917fc68645dc6fa8bf804">Chromosome::GetRawGene</a>). </p><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">geneBitWidth</td><td>The width of the gene we want to add. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="classpanga_1_1Chromosome.html#abe6e9d64ff7917fc68645dc6fa8bf804">Chromosome::GetRawGene</a> </dd></dl>
<dl class="section return"><dt>Returns</dt><dd>The gene index of the newly-added gene. </dd></dl>

</div>
</div>
<a id="a54c4cb4c89662ddd701716235147ba5e"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a54c4cb4c89662ddd701716235147ba5e">&#9670;&nbsp;</a></span>BitsRequired()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">size_t Genome::BitsRequired </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Get the number of bits needed to encode this <a class="el" href="classpanga_1_1Genome.html">Genome</a>. </p>

</div>
</div>
<a id="a8a2f7f6c10a0e96addd4a71a84d699cf"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a8a2f7f6c10a0e96addd4a71a84d699cf">&#9670;&nbsp;</a></span>GetBooleanGeneCount()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">size_t Genome::GetBooleanGeneCount </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Get the number of boolean Genes in the <a class="el" href="classpanga_1_1Genome.html">Genome</a>. </p>

</div>
</div>
<a id="a56f745762146d7579823b54d4acdbc9f"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a56f745762146d7579823b54d4acdbc9f">&#9670;&nbsp;</a></span>GetFirstBooleanGeneBitIndex()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">size_t Genome::GetFirstBooleanGeneBitIndex </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Get the bit index of the first boolean gene in the <a class="el" href="classpanga_1_1Genome.html">Genome</a>. </p>

</div>
</div>
<a id="af5432ac15e82a2c8693affa88375b202"></a>
<h2 class="memtitle"><span class="permalink"><a href="#af5432ac15e82a2c8693affa88375b202">&#9670;&nbsp;</a></span>GetFirstBooleanGeneIndex()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">size_t Genome::GetFirstBooleanGeneIndex </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Get the gene index of the first boolean gene in the <a class="el" href="classpanga_1_1Genome.html">Genome</a>. </p>

</div>
</div>
<a id="ac9db69ece5360c5eaa94c9cbc8b5d0b6"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ac9db69ece5360c5eaa94c9cbc8b5d0b6">&#9670;&nbsp;</a></span>GetGeneBitWitdh()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">size_t Genome::GetGeneBitWitdh </td>
          <td>(</td>
          <td class="paramtype">size_t&#160;</td>
          <td class="paramname"><em>geneIndex</em></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Get the bit width for a gene. </p><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">geneIndex</td><td>Index of the gene we should return the width for. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>The width in bits of gene at index geneIndex. </dd></dl>

</div>
</div>
<a id="a0440fd8216d128563f22e1d4e1d17303"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a0440fd8216d128563f22e1d4e1d17303">&#9670;&nbsp;</a></span>GetGeneCount()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">size_t Genome::GetGeneCount </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Get the count of genes in the <a class="el" href="classpanga_1_1Genome.html">Genome</a>.<br />
 Note: Includes the count of boolean genes. </p>

</div>
</div>
<a id="a81ef777300e2c0861c69de1421baab8c"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a81ef777300e2c0861c69de1421baab8c">&#9670;&nbsp;</a></span>GetGeneStartBitIndex()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">size_t Genome::GetGeneStartBitIndex </td>
          <td>(</td>
          <td class="paramtype">size_t&#160;</td>
          <td class="paramname"><em>geneIndex</em></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Get the bit offset where a gene starts. </p>

</div>
</div>
<a id="adf73367588c8932a4379e9d2d92d02ff"></a>
<h2 class="memtitle"><span class="permalink"><a href="#adf73367588c8932a4379e9d2d92d02ff">&#9670;&nbsp;</a></span>SetBooleanGeneCount()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void Genome::SetBooleanGeneCount </td>
          <td>(</td>
          <td class="paramtype">size_t&#160;</td>
          <td class="paramname"><em>booleanGeneCount</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Set the number of boolean genes in this <a class="el" href="classpanga_1_1Genome.html">Genome</a>.<br />
 This overwrites any value currently in the boolean gene count. </p>

</div>
</div>
<hr/>The documentation for this class was generated from the following files:<ul>
<li>src/<a class="el" href="Genome_8h_source.html">Genome.h</a></li>
<li>src/Genome.cc</li>
</ul>
</div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by &#160;<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/>
</a> 1.8.13
</small></address>
</body>
</html>
